<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BeeBDC">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Basic workflow • BeeBDC</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Basic workflow">
<meta property="og:description" content="BeeBDC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BeeBDC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--articles">
    <span class="fas fa fas fa-book"></span>
     
    Articles
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--articles">
    <a class="dropdown-item" href="../articles/BeeBDC_main.html">BeeBDC vignette</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#data-merge">  1.0 Data merge</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#data-preparation">  2.0 Data preparation</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#initial-flags">  3.0 Initial flags</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#taxonomy">  4.0 Taxonomy</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#space">  5.0 Space</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#time">  6.0 Time</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#de-duplication">  7.0 De-duplication</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#data-filtering">  8.0 Data filtering</a>
    <a class="dropdown-item" href="../articles/BeeBDC_main.html#figures-and-tables">  9.0 Figures and tables</a>
    <a class="dropdown-item" href="../articles/basic_workflow.html">Basic workflow</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jbdorey/BeeBDC/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        
        
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Basic workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jbdorey/BeeBDC/blob/HEAD/vignettes/basic_workflow.Rmd" class="external-link"><code>vignettes/basic_workflow.Rmd</code></a></small>
      <div class="d-none name"><code>basic_workflow.Rmd</code></div>
    </div>

    
    
<p>This workflow is meant to be a basic example workflow of how a user
might take a flagged version of our (or some other) occurrence dataset
and filter for specific taxa or countries, re-apply flagging functions,
re-filter the data, or make maps based of those data.</p>
<div class="section level2">
<h2 id="script-preparation">0.0 Script preparation<a class="anchor" aria-label="anchor" href="#script-preparation"></a>
</h2>
<div class="section level3">
<h3 id="working-directory">0.1 Working directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>Choose the path to the root folder in which all other folders can be
found.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RootPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"/your/path/here"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Create the working directory in the RootPath if it doesn't exist already</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>, <span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>, <span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>  <span class="co"># Set the working directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>,<span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For the first time that you run BeeBDC, and if you want to use the
renv package to manage your packages, you can install renv…</p>
<pre><code><span>        <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"renv"</span>, repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span></code></pre>
<p>and then initialise renv the project.</p>
<pre><code><span>        <span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/init.html" class="external-link">init</a></span><span class="op">(</span>project <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>,<span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
<p>If you have already initialised a project, you can instead just
activate it.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span>project <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>,<span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-packages-if-needed">0.2 Install packages (if needed)<a class="anchor" aria-label="anchor" href="#install-packages-if-needed"></a>
</h3>
<p>You may need to install gdal on your computer. This can be done on a
Mac by using Homebrew in the terminal and the command “brew install
gdal”.</p>
<p>To start out, you will need to install <strong>BiocManager</strong>,
<strong>devtools</strong>, <strong>ComplexHeatmap</strong>, and
<strong>rnaturalearthhires</strong> to then install and fully use
<strong>BeeBDC</strong>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Install remotes if needed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span>, repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span>  <span class="co"># Download and then load rnaturalearthhires</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ropensci/rnaturalearthhires"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rnaturalearthhires"</span>, repos <span class="op">=</span> <span class="st">"https://ropensci.r-universe.dev"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/rnaturalearthhires" class="external-link">rnaturalearthhires</a></span><span class="op">)</span></span></code></pre></div>
<p>Now install <strong>BeeBDC</strong>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/jbdorey/BeeBDC.git"</span>, ref <span class="op">=</span> <span class="st">"main"</span>, </span>
<span>                        force <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BeeBDC</span><span class="op">)</span></span></code></pre></div>
<p>Snapshot the renv environment.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/snapshot.html" class="external-link">snapshot</a></span><span class="op">(</span>project <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>,<span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span>,</span>
<span>                 prompt <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">## The following package(s) will be updated in the lockfile:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## # RSPM -----------------------------------------------------------------------</span></span>
<span><span class="co">## - renv   [* -&gt; 1.0.2]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The version of R recorded in the lockfile will be updated:</span></span>
<span><span class="co">## - R      [* -&gt; 4.3.1]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## - Lockfile written to "/tmp/RtmpoTgAdZ/Data_acquisition_workflow/renv.lock".</span></span></code></pre></div>
<p>Set up the directories used by <strong>BeeBDC</strong>. These
directories include where the data, figures, reports, etc. will be
saved. The RDoc needs to be a path RELATIVE to the RootPath; i.e., the
file path from which the two diverge.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dirMaker.html">dirMaker</a></span><span class="op">(</span></span>
<span>    RootPath <span class="op">=</span> <span class="va">RootPath</span>,</span>
<span>    RDoc <span class="op">=</span> <span class="st">"vignettes/BeeBDC_main.Rmd"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="co"># Add paths created by this function to the environment()</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list2env.html" class="external-link">list2env</a></span><span class="op">(</span>envir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">environment</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-packages">0.3 Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>Load packages.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span>, <span class="st">"magrittr"</span><span class="op">)</span>, </span>
<span>       <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## Loading required package: grid</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'grid'</span></span>
<span><span class="co">## The following object is masked from 'package:terra':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     depth</span></span>
<span><span class="co">## ========================================</span></span>
<span><span class="co">## ComplexHeatmap version 2.16.0</span></span>
<span><span class="co">## Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/</span></span>
<span><span class="co">## Github page: https://github.com/jokergoo/ComplexHeatmap</span></span>
<span><span class="co">## Documentation: http://jokergoo.github.io/ComplexHeatmap-reference</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## If you use it in published research, please cite either one:</span></span>
<span><span class="co">## - Gu, Z. Complex Heatmap Visualization. iMeta 2022.</span></span>
<span><span class="co">## - Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional </span></span>
<span><span class="co">##     genomic data. Bioinformatics 2016.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The new InteractiveComplexHeatmap package can directly export static </span></span>
<span><span class="co">## complex heatmaps into an interactive Shiny app with zero effort. Have a try!</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## This message can be suppressed by:</span></span>
<span><span class="co">##   suppressPackageStartupMessages(library(ComplexHeatmap))</span></span>
<span><span class="co">## ========================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ComplexHeatmap'</span></span>
<span><span class="co">## The following object is masked from 'package:terra':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     draw</span></span>
<span><span class="co">## The following object is masked from 'package:R.utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     draw</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="taxon-example">2.0 Taxon example<a class="anchor" aria-label="anchor" href="#taxon-example"></a>
</h2>
<p>If you want to filter the dataset to a particular taxon of interest,
you can do so quite easily using <strong>dplyr</strong> from the
<strong>tidyverse</strong> group of packages. To filter to a selected
bee genus, in our case Anthophorini…</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load some package data — the taxonomy and a flagged example dataset</span></span>
<span>  <span class="co"># Download the full beesTaxonomy file</span></span>
<span><span class="va">beesTaxonomy</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/beesTaxonomy.html">beesTaxonomy</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: 'beesTaxonomy' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Load the example beesFlagged dataset</span></span>
<span><span class="va">beesFlagged</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="va"><a href="../reference/beesFlagged.html">beesFlagged</a></span></span>
<span></span>
<span><span class="va">selectedGenera</span> <span class="op">&lt;-</span> <span class="va">beesTaxonomy</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Select only tribe anthophorini (for example)</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">tribe</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="st">"anthophorini"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">genus</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in eval(expr, envir, enclos): object 'beesTaxonomy' not found</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Filter the data</span></span>
<span><span class="va">taxonData</span> <span class="op">&lt;-</span> <span class="va">beesFlagged</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genus</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selectedGenera</span><span class="op">$</span><span class="va">genus</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `dplyr::filter()`:</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> In argument: `genus %in% selectedGenera$genus`.</span></span>
<span><span class="co">## <span style="font-weight: bold;">Caused by error in `h()`:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> error in evaluating the argument 'table' in selecting a method for function '%in%': object 'selectedGenera' not found</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># View the data</span></span>
<span><span class="va">taxonData</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in eval(expr, envir, enclos): object 'taxonData' not found</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="country-example">3.0 Country example<a class="anchor" aria-label="anchor" href="#country-example"></a>
</h2>
<p>Similarly to the above you can filter for only countries of interest.
Keep in mind, that sometimes the <em>country</em> column may not hold
all of the records that fall in that country, if it, or the coordinates,
have been entered incorrectly.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Select your study area</span></span>
<span><span class="va">studyArea</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Canada"</span>, <span class="st">"United states"</span>, <span class="st">"Mexico"</span>, <span class="st">"Guatemala"</span><span class="op">)</span></span>
<span><span class="co"># Filter the data to that area</span></span>
<span><span class="va">countryData</span> <span class="op">&lt;-</span> <span class="va">beesFlagged</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">country</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">studyArea</span><span class="op">)</span></span>
<span>  <span class="co"># View the data</span></span>
<span><span class="va">countryData</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 49 × 124</span></span></span>
<span><span class="co">##    database_id scientificName family subfamily genus subgenus subspecies species</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Dorey_data… Macrotera arc… Andre… Panurgin… Macr… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Macrot…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Dorey_data… Exomalopsis s… Apidae Apinae    Exom… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Exomal…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Paige_data… Augochlorella… Halic… Halictin… Augo… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Augoch…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Dorey_data… Svastra duplo… Apidae Apinae    Svas… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Svastr…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Dorey_data… Agapostemon f… Halic… Halictin… Agap… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Agapos…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Dorey_data… Melissodes tr… Apidae Apinae    Meli… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Meliss…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Dorey_data… Osmia pumila … Megac… Megachil… Osmia <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Osmia …</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Dorey_data… Perdita bisho… Andre… Panurgin… Perd… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Perdit…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Dorey_data… Melissodes lu… Apidae Apinae    Meli… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Meliss…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Paige_data… Melecta thora… Apidae Apinae    Mele… <span style="color: #BB0000;">NA</span>       <span style="color: #BB0000;">NA</span>         Melect…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 39 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 116 more variables: specificEpithet &lt;chr&gt;, infraspecificEpithet &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   acceptedNameUsage &lt;lgl&gt;, taxonRank &lt;chr&gt;, scientificNameAuthorship &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   identificationQualifier &lt;lgl&gt;, higherClassification &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   identificationReferences &lt;lgl&gt;, typeStatus &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   previousIdentifications &lt;chr&gt;, verbatimIdentification &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   identifiedBy &lt;chr&gt;, dateIdentified &lt;chr&gt;, decimalLatitude &lt;dbl&gt;, …</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="filtering-example">4.0 Filtering example<a class="anchor" aria-label="anchor" href="#filtering-example"></a>
</h2>
<div class="section level3">
<h3 id="simple-filter">4.1 Simple filter<a class="anchor" aria-label="anchor" href="#simple-filter"></a>
</h3>
<p>The <strong>BeeBDC</strong> package provides a simple function that
can re-build the <em>.summary</em> column based off of the filtering
columns that are present in the dataset (those starting with “.”). you
can also choose which filters you DO NOT want to implement using the
dontFilterThese argument. In this example, we are also removing all of
the filtering columns in the output dataset (removeFilterColumns = TRUE)
and filtering to only completely clean occurrences (filterClean = TRUE).
For the latter, we are only keeping <em>.summary</em> == TRUE.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filteredData</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">beesFlagged</span>,</span>
<span>   <span class="co"># Choose the columns to NOT filter (or NULL to filter all columns)</span></span>
<span>   dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>, <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span>,</span>
<span>                      <span class="st">".uncertaintyThreshold"</span><span class="op">)</span>,</span>
<span>    <span class="co"># In the output, do you want to REMOVE all filtering columns (TRUE), or keep them (FALSE)</span></span>
<span>   removeFilterColumns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>   <span class="co"># In the output, do you want to only keep clean data according to your filtering (TRUE),</span></span>
<span>    <span class="co"># Or keep all data and simply update the .summary column (FALSE)</span></span>
<span>  filterClean <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Error: 'summaryFun' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="uncertainty-threshold">4.2 Uncertainty threshold<a class="anchor" aria-label="anchor" href="#uncertainty-threshold"></a>
</h3>
<p>You may also want to change the <em>.uncertaintyThreshold</em> as we
have chosen a somewhat strict default of 1 km in our dataset. Here, we
will instead flag to 10 km (threshold = 10000 [m]). Additionally, we use
the <strong>magrittr</strong> package pipe (%&gt;%) to feed the outputs
directly into <code><a href="../reference/summaryFun.html">summaryFun()</a></code> to filter our data in one
action!</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filteredData</span> <span class="op">&lt;-</span> <span class="va">beesFlagged</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Remove any exiting .uncertaintyThreshold column</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">any_of</a></span><span class="op">(</span><span class="st">".uncertaintyThreshold"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Chose the coordinate uncertainty to filter to...</span></span>
<span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/coordUncerFlagR.html">coordUncerFlagR</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">.</span>,</span>
<span>                  uncerColumn <span class="op">=</span> <span class="st">"coordinateUncertaintyInMeters"</span>,</span>
<span>                    <span class="co"># 10 km here</span></span>
<span>                  threshold <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Now re-do the .summary column and filter the data using this new value</span></span>
<span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">.</span>,</span>
<span>  dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>, <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span><span class="op">)</span>,</span>
<span>  removeFilterColumns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterClean <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: 'summaryFun' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="date-filter">4.2 Date filter<a class="anchor" aria-label="anchor" href="#date-filter"></a>
</h3>
<div class="section level4">
<h4 id="a--bdc_year_outofrange">a. bdc_year_outOfRange<a class="anchor" aria-label="anchor" href="#a--bdc_year_outofrange"></a>
</h4>
<p>Another column that users are likely to want to pay close attention
to is the <em>.year_outOfRange</em> column that is set at 1950 in our
dataset. In this case, <strong>bdc</strong> provides the function where
users can change the year_threshold argument to, in this case, 1970. As
with above, we then use <code><a href="../reference/summaryFun.html">summaryFun()</a></code> to get results in one
go.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filteredData</span> <span class="op">&lt;-</span> <span class="va">beesFlagged</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Remove any exisitng .year_outOfRange column</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="st">".year_outOfRange"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Chose the minimum year to filter to...</span></span>
<span>  <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_year_outOfRange.html" class="external-link">bdc_year_outOfRange</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">.</span>,</span>
<span>                           eventDate <span class="op">=</span> <span class="st">"year"</span>,</span>
<span>                           year_threshold <span class="op">=</span> <span class="fl">1970</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Now re-do the .summary column and filter the data using this new value</span></span>
<span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">.</span>,</span>
<span>    dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>, <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span>,</span>
<span>                        <span class="st">".uncertaintyThreshold"</span><span class="op">)</span>,</span>
<span>    removeFilterColumns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    filterClean <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: 'summaryFun' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="b--year-range">b. year range<a class="anchor" aria-label="anchor" href="#b--year-range"></a>
</h4>
<p>Or, if you’re interested in a particular time period, again
<strong>dplyr</strong> comes to the rescue with some very straight
forward filtering within a year range.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filteredData</span> <span class="op">&lt;-</span> </span>
<span>  <span class="co"># The input dataset</span></span>
<span>  <span class="va">beesFlagged</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Chose the year range...</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">year</span> <span class="op">&gt;</span> <span class="fl">1950</span> <span class="op">&amp;</span> <span class="va">year</span> <span class="op">&lt;</span> <span class="fl">1970</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Now re-do the .summary column and filter the data using this new value</span></span>
<span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span></span>
<span>    <span class="co"># Select the input dataset to filter</span></span>
<span>    data <span class="op">=</span> <span class="va">.</span>,</span>
<span>    <span class="co"># Choose the columns to NOT filter (or NULL to filter all columns)</span></span>
<span>    dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>, <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span>,</span>
<span>                        <span class="st">".uncertaintyThreshold"</span><span class="op">)</span>,</span>
<span>    <span class="co"># In the output, do you want to REMOVE all filtering columns (TRUE), or keep them (FALSE)</span></span>
<span>    removeFilterColumns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    <span class="co"># In the output, do you want to only keep clean data according to your filtering (TRUE),</span></span>
<span>    <span class="co"># Or keep all data and simply update the .summary column (FALSE)</span></span>
<span>    filterClean <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: 'summaryFun' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
<p>Users may choose any number of filtering steps form the main workflow
to include above <code><a href="../reference/summaryFun.html">summaryFun()</a></code>, just use pipes ‘%&gt;%’
between the function and use ‘.’ as the data input because this will
feed in the data aoutput from the above function into the proceeding
one.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="summary-figures">5. Summary figures<a class="anchor" aria-label="anchor" href="#summary-figures"></a>
</h2>
<p>Now, if you wanted to rebuild some figures, say after you’ve added or
filtered data, then you can use some of the below processes.</p>
<div class="section level3">
<h3 id="duplicate-chorddiagrams">5.1 Duplicate chordDiagrams<a class="anchor" aria-label="anchor" href="#duplicate-chorddiagrams"></a>
</h3>
<p>Our <code><a href="../reference/chordDiagramR.html">chordDiagramR()</a></code> function is very useful and it
relies on two great packages, <strong>circlize</strong> and
<strong>ComplexHeatmap</strong>. Unfortunately, the latter is not
available on CRAN and so must be downloaded using
<strong>BiocManager</strong>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span>, force <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/snapshot.html" class="external-link">snapshot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We don’t actually have an example duplicates dataset with the
package, so I’ll magic one up behind the scences!</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">duplicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fileFinder.html">fileFinder</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"PATH TO A FOLDER CONTAINING THE duplicateRun_ — could be supp. materials folder"</span>,</span>
<span>                          fileName <span class="op">=</span> <span class="st">"duplicateRun_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Select only the stingless bee data</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">database_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">stinglessData</span><span class="op">$</span><span class="va">database_id</span> <span class="op">|</span></span>
<span>                  <span class="va">database_id_match</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">stinglessData</span><span class="op">$</span><span class="va">database_id</span><span class="op">)</span></span></code></pre></div>
<p>Then, set some parameters for figure borders and run your data
through <code><a href="../reference/chordDiagramR.html">chordDiagramR()</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Choose the global figure parameters</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the chorDiagram. You can leave many of the below values out but we show here</span></span>
<span><span class="co"># the defaults</span></span>
<span></span>
<span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/chordDiagramR.html">chordDiagramR</a></span><span class="op">(</span></span>
<span>  <span class="co"># The duplicate data from the dupeSummary function output  </span></span>
<span>  dupeData <span class="op">=</span> <span class="va">duplicates</span>,</span>
<span>  outPath <span class="op">=</span> <span class="va">OutPath_Figures</span>,</span>
<span>  fileName <span class="op">=</span> <span class="st">"ChordDiagram.pdf"</span>,</span>
<span>  <span class="co"># These can be modified to help fit the final pdf that's exported.</span></span>
<span>  width <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">7.5</span>,</span>
<span>  bg <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  <span class="co"># How few distinct dataSources should a group have to be listed as "other"</span></span>
<span>  smallGrpThreshold <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Duplicated record sources"</span>,</span>
<span>  <span class="co"># The default list of colour palettes to choose from usign the paleteer package</span></span>
<span>  palettes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cartography::blue.pal"</span>, <span class="st">"cartography::green.pal"</span>, </span>
<span>               <span class="st">"cartography::sand.pal"</span>, <span class="st">"cartography::orange.pal"</span>, <span class="st">"cartography::red.pal"</span>,</span>
<span>               <span class="st">"cartography::purple.pal"</span>, <span class="st">"cartography::brown.pal"</span><span class="op">)</span>,</span>
<span>  canvas.ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.0</span>,<span class="fl">1.0</span><span class="op">)</span>, </span>
<span>  canvas.xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.6</span>, <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>  text.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  legendX <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  legendY <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">18</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  legendJustify <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span>,</span>
<span>  niceFacing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="duplicate-histogram">5.2 Duplicate histogram<a class="anchor" aria-label="anchor" href="#duplicate-histogram"></a>
</h3>
<p>In this example, we will use one of the example datasets to show you
how this works. We will use beesFlagged, which has been filtered from a
larger dataset and contains duplicates from that larger dataset. To
print the plot in R, you need to specify returnPlot = TRUE, otherwise it
will only save to the disk</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"beesFlagged"</span>, package <span class="op">=</span> <span class="st">"BeeBDC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a figure shoring the total number of duplicates, kept duplicates, and unique</span></span>
<span><span class="co"># records for each datasource (simplified to the text before the first underscore) and</span></span>
<span><span class="co"># the proportion of the above for each data source</span></span>
<span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dupePlotR.html">dupePlotR</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">beesFlagged</span>,</span>
<span>  <span class="co"># The outPath to save the plot as</span></span>
<span>  outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  fileName <span class="op">=</span> <span class="st">"Fig3_duplicatePlot.pdf"</span>,</span>
<span>  <span class="co"># Colours in order: duplicate, kept duplicate, unique</span></span>
<span>  dupeColours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F2D2A2"</span>,<span class="st">"#B9D6BC"</span>, <span class="st">"#349B90"</span><span class="op">)</span>,</span>
<span>  <span class="co"># Plot size and height</span></span>
<span>  base_height <span class="op">=</span> <span class="fl">7</span>, base_width <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.85</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>  <span class="co"># Extra variables can be fed into forcats::fct_recode() to change names on plot</span></span>
<span>  GBIF <span class="op">=</span> <span class="st">"GBIF"</span>, SCAN <span class="op">=</span> <span class="st">"SCAN"</span>, iDigBio <span class="op">=</span> <span class="st">"iDigBio"</span>, USGS <span class="op">=</span> <span class="st">"USGS"</span>, ALA <span class="op">=</span> <span class="st">"ALA"</span>, </span>
<span>  ASP <span class="op">=</span> <span class="st">"ASP"</span>,</span>
<span>  returnPlot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: 'dupePlotR' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="flags-by-source">5.3 Flags by source<a class="anchor" aria-label="anchor" href="#flags-by-source"></a>
</h3>
<p>The <code><a href="../reference/plotFlagSummary.html">plotFlagSummary()</a></code> function is one of the most
important for quickly summarising and checking that your data and flags
have worked together correctly. It can be a good starting point for
error-checking. You will also see in <code><a href="../reference/plotFlagSummary.html">plotFlagSummary()</a></code> that
you can filter to particular species and also output quick point maps of
those species.</p>
<div class="section level4">
<h4 id="a--all-taxa-in-dataset">a. all taxa in dataset<a class="anchor" aria-label="anchor" href="#a--all-taxa-in-dataset"></a>
</h4>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualise-all-flags-for-each-datasource-simplified-to-the-text-before-the-first-underscore">Visualise all flags for each dataSource (simplified to the text
before the first underscore)<a class="anchor" aria-label="anchor" href="#visualise-all-flags-for-each-datasource-simplified-to-the-text-before-the-first-underscore"></a>
</h2>
<p>BeeBDC::plotFlagSummary( data = beesFlagged, # Colours in order of
pass (TRUE), fail (FALSE), and NA flagColours = c(“#127852”, “#A7002D”,
“#BDBABB”), fileName = paste0(“Fig4_FlagsPlot_”, Sys.Date(),“.pdf”),
outPath = tempdir(), width = 15, height = 9, # Extra variables can be
fed into forcats::fct_recode() to change names on plot GBIF = “GBIF”,
SCAN = “SCAN”, iDigBio = “iDigBio”, USGS = “USGS”, ALA = “ALA”, ASP =
“ASP”, returnPlot = TRUE )</p>
<pre><code>

  ###### b. Single sp. summary ####
  
  In fact, lets build one of these single-species example below using the same data and the omnipresent *Apis mellifera*.
  

```r
# Visualise all flags for each dataSource (simplified to the text before the first underscore)
  # A clever user might also realise the potential to summarise and produce outputs in other columns
BeeBDC::plotFlagSummary(
  # WARNING: alternate path if wanting to produce figures for the selected taxonData (2.0 above)
  # Select only the taxonData data
  data = beesFlagged,
  # Colours in order of pass (TRUE), fail (FALSE), and NA
  flagColours = c("#127852", "#A7002D", "#BDBABB"),
  fileName = paste0("FlagsPlot_Amell", Sys.Date(),".pdf"),
  outPath = tempdir(),
  width = 15, height = 9,
  # OPTIONAL:
         #  # Filter to species
           speciesName = "Apis mellifera Linnaeus, 1758",
             # column to look in
           nameColumn = "scientificName",
           # Save the filtered data
           saveFiltered = FALSE,
     # Filter column to display on map
           filterColumn = ".summary",
           plotMap = TRUE,
       # amount to jitter points if desired, e.g. 0.25 or NULL
     jitterValue = NULL,
       # Map opacity value for points between 0 and 1
     mapAlpha = 1,
  returnPlot = TRUE,
  # Extra variables can be fed into forcats::fct_recode() to change names on plot
  GBIF = "GBIF", SCAN = "SCAN", iDigBio = "iDigBio", USGS = "USGS", ALA = "ALA", 
  ASP = "ASP", CAES = "CAES", 'B. Mont.' = "BMont", 'B. Minkley' = "BMin", Ecd = "Ecd",
  Gaiarsa = "Gai", EPEL = "EPEL"
)</code></pre>
<pre><code><span><span class="co">## Error: 'plotFlagSummary' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
<div class="section level3">
<h3 id="maps">5.4 Maps<a class="anchor" aria-label="anchor" href="#maps"></a>
</h3>
<p>We can also make some overall summary maps at the country level using
<code><a href="../reference/summaryMaps.html">summaryMaps()</a></code>. If you get an error about breaks not being
unique, then reduce class_n.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryMaps.html">summaryMaps</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">beesFlagged</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">10</span>, height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  class_n <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  class_Style <span class="op">=</span> <span class="st">"jenks"</span>,</span>
<span>  outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  fileName <span class="op">=</span> <span class="st">"CountryMaps_jenks.pdf"</span>,</span>
<span>  returnPlot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error: 'summaryMaps' is not an exported object from 'namespace:BeeBDC'</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="save-data">6.0 Save data<a class="anchor" aria-label="anchor" href="#save-data"></a>
</h2>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapData</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">DataPath</span>, <span class="st">"/Output/Intermediate/"</span>, <span class="st">"cleanTaxon_"</span>,</span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">".csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by James B. Dorey, Robert L. O’Reilly, Silas Bossert, Erica E. Fischer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
