<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-gb">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>BeeBDC vignette • BeeBDC</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="BeeBDC vignette">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BeeBDC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html" aria-label="Function list"><span class="fa fas fa-r"></span> Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-occurrence-cleaning" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Occurrence data cleaning vignettes"><span class="fa fas fa-broom"></span> Occurrence cleaning</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-occurrence-cleaning">
<li><a class="dropdown-item" href="../articles/BeeBDC_main.html">BeeBDC full cleaning workflow (0.0–9.0)</a></li>
    <li><a class="dropdown-item" href="../articles/basic_workflow.html">BeeBDC basic workflow</a></li>
    <li><a class="dropdown-item" href="../articles/dataPrep.html">Bee-specific data preparation</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-species-richness-estimation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Species richness estimation vignettes"><span class="fa fas fa-arrow-up-right-dots"></span> Species richness estimation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-species-richness-estimation">
<li><a class="dropdown-item" href="../articles/speciesRichness_example.html">BeeBDC richness estimation</a></li>
    <li><a class="dropdown-item" href="../articles/speciesRichness_example.html#Parallel-estimations">  1.0 Parallel estimation</a></li>
    <li><a class="dropdown-item" href="../articles/speciesRichness_example.html#Estimating-richness-from-the-known-unknowns">  2.0 Estimation with checklists</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="https://jbdorey.github.io/BeeBDC/news/index.html"><span class="fa fas fa-newspaper"></span> News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jbdorey/BeeBDC/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>BeeBDC vignette</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jbdorey/BeeBDC/blob/main/vignettes/articles/BeeBDC_main.Rmd" class="external-link"><code>vignettes/articles/BeeBDC_main.Rmd</code></a></small>
      <div class="d-none name"><code>BeeBDC_main.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="section">
<a href="https://github.com/jbdorey/BeeBDC" class="external-link"><img src="https://photos.smugmug.com/photos/i-MpLFKTT/0/741daa6d/X4/i-MpLFKTT-X4.png" alt="BeeBDC logo of a cuckoo bee sweeping up occurrence records in South America" align="right" width="155"></a><a class="anchor" aria-label="anchor" href="#section"></a>
</h2>
</div>
<div class="section level2">
<h2 id="script-preparation">0.0 Script preparation<a class="anchor" aria-label="anchor" href="#script-preparation"></a>
</h2>
<div class="section level3">
<h3 id="working-directory">0.1 Working directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>To start off with, have a think about where you want to work from.
<strong>BeeBDC</strong> and <strong>bdc</strong> can create quite a few
files and so setting this up well from the start is a good idea. If you
are afraid that you might run out of storage, this could also be on a
hard drive; but you can always change that later. Defining your RootPath
at the top of your script, only once, shoudl make your life easier.</p>
<p>Choose the path to the root folder in which all other folders can be
found.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RootPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"/your/path/here"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create the working directory in the RootPath if it doesn't exist already</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>, <span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>, <span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Set the working directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">RootPath</span>, <span class="st">"/Data_acquisition_workflow"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-packages-if-needed">0.2 Install packages (if needed)<a class="anchor" aria-label="anchor" href="#install-packages-if-needed"></a>
</h3>
<blockquote>
<h4 id="is-this-your-first-time-using-the-sf-or-terra-packages">Is this
your first time using the <em>sf</em> or <em>terra</em> packages?</h4>
<p>The first time that you use <strong>terra</strong> or
<strong>sf</strong> on a new computer you may need to install some
dependencies. Try to install the <strong>terra</strong> and
<strong>sf</strong> packages first but then come back here if that
doesn’t work.</p>
<h6 id="windows">Windows:</h6>
<p>On <strong>Windows</strong>, you need to first install <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a> to get
a C++ compiler that <strong>R</strong> can use. You need a recent
version of <strong>Rtools42</strong> (rtools42-5355-5357).</p>
<h6 id="macos">MacOS:</h6>
<p>On <strong>macOS</strong>, you can use <a href="https://www.macports.org/" class="external-link">MacPorts</a> or <a href="https://brew.sh/" class="external-link">Homebrew</a>.</p>
<p>With <strong>MacPorts</strong> you can do</p>
<p><code>sudo port install R-terra</code></p>
<p>With <strong>Homebrew</strong>, you need to first install GDAL:</p>
<p><code>brew install pkg-config</code></p>
<p><code>brew install gdal</code></p>
<p>Followed by (note the additional configuration argument needed for
Homebrew)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install terra</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"terra"</span>, type <span class="op">=</span> <span class="st">"source"</span>, configure.args <span class="op">=</span> <span class="st">"--with-proj-lib=$(brew --prefix)/lib/"</span><span class="op">)</span></span>
<span><span class="co"># install sf</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"sf"</span>, type <span class="op">=</span> <span class="st">"source"</span>, configure.args <span class="op">=</span> <span class="st">"--with-proj-lib=$(brew --prefix)/lib/"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span></code></pre></div>
</blockquote>
<p>If you have <strong>sf</strong> and <strong>terra</strong> isntalled,
you can now install <strong>BeeBDC</strong>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BeeBDC"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jbdorey.github.io/BeeBDC/">BeeBDC</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>BeeBDC</strong> also has a few optional packages that are
required for a subset of the functions. You don’t need to isntall these
now, if you don’t want to, but you can do so later!</p>
<p>You can optionally install <strong>BiocManager</strong>,
<strong>devtools</strong>, <strong>ComplexHeatmap</strong>,
<strong>rnaturalearthhires</strong>, and <strong>taxadb</strong> or do
it later as you wish.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install remotes if needed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span>, repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="co"># Download and then load rnaturalearthhires</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ropensci/rnaturalearthhires"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rnaturalearthhires"</span>, repos <span class="op">=</span> <span class="st">"https://ropensci.r-universe.dev"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rnaturalearthhires</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"taxadb"</span><span class="op">)</span></span></code></pre></div>
<p>Set up the directories used by <strong>BeeBDC</strong>. These
directories include where the data, figures, reports, etc. will be
saved. The RDoc needs to be a path RELATIVE to the RootPath; i.e., the
file path from which the two diverge.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dirMaker.html">dirMaker</a></span><span class="op">(</span>RootPath <span class="op">=</span> <span class="va">RootPath</span>, RDoc <span class="op">=</span> <span class="st">"vignettes/BeeBDC_main.Rmd"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Add paths created by this function to the environment()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list2env.html" class="external-link">list2env</a></span><span class="op">(</span>envir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">parent.env</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">environment</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-packages">0.3 Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>Let’s go ahead and load our packages before we start!</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span>, <span class="st">"magrittr"</span><span class="op">)</span>, <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## Loading required package: grid</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'grid'</span></span>
<span><span class="co">## The following object is masked from 'package:terra':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     depth</span></span>
<span><span class="co">## ========================================</span></span>
<span><span class="co">## ComplexHeatmap version 2.24.1</span></span>
<span><span class="co">## Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/</span></span>
<span><span class="co">## Github page: https://github.com/jokergoo/ComplexHeatmap</span></span>
<span><span class="co">## Documentation: http://jokergoo.github.io/ComplexHeatmap-reference</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## If you use it in published research, please cite either one:</span></span>
<span><span class="co">## - Gu, Z. Complex Heatmap Visualization. iMeta 2022.</span></span>
<span><span class="co">## - Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional </span></span>
<span><span class="co">##     genomic data. Bioinformatics 2016.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The new InteractiveComplexHeatmap package can directly export static </span></span>
<span><span class="co">## complex heatmaps into an interactive Shiny app with zero effort. Have a try!</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## This message can be suppressed by:</span></span>
<span><span class="co">##   suppressPackageStartupMessages(library(ComplexHeatmap))</span></span>
<span><span class="co">## ========================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ComplexHeatmap'</span></span>
<span><span class="co">## The following object is masked from 'package:terra':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     draw</span></span>
<span><span class="co">## The following object is masked from 'package:R.utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     draw</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="data-merge">1.0 Data merge<a class="anchor" aria-label="anchor" href="#data-merge"></a>
</h2>
<p>There are many ways to get a good starting dataset. For
<strong>bees</strong>, we have compiled a bunch of large datasets that
we will periodically update and make available. However, we have also
made available the code to update many of these public datasets yourself
in the <a href="articles/dataPrep.html"><strong>bee data prep
vignette</strong></a>.</p>
<p>If you are looking to just make your own dataset to feed in, for
example, from another taxon. Great! We have some usefull tools below and
so does the <strong>bdc</strong> package! Feel free to take you pick at
how you compile your starting data but just note that
<strong>BeeBDC</strong>, <strong>bdc</strong>, and
<strong>coordinateCleaner</strong> like (and often assume) that your
data are in a <a href="https://dwc.tdwg.org" class="external-link">Darwin Core</a> format.</p>
<div class="alert alert-info">
<p><strong> Attention:</strong> <br> Although each line of code has been
validated, in order to save time knitting the R
<strong>markdown</strong> document the next section is display only. If
you are not data merging (section 1.0) or preparing the data (section
2.0), feel free to skip to Section 3.0 Initial flags.</p>
</div>
<div class="section level3">
<h3 id="download-ala-data">1.1 Download ALA data<a class="anchor" aria-label="anchor" href="#download-ala-data"></a>
</h3>
<p>If you’re interested in using data from the <em>Atlas of Living
Australia (ALA)</em> or one of the other Atlas repositories, you can use
<code><a href="../reference/atlasDownloader.html">BeeBDC::atlasDownloader()</a></code> below to access and download
those files and their metadata easily enough. You may, however, need to
ensure that you have an account that you can link to the download.
Especially fo the sake of a doi.</p>
<p>To make an account with ALA in order to download your data visit this
link — <a href="https://auth.ala.org.au/userdetails/registration/createAccount" class="external-link uri">https://auth.ala.org.au/userdetails/registration/createAccount</a></p>
<pre><code><span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/atlasDownloader.html">atlasDownloader</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>,</span>
<span>           userEmail <span class="op">=</span> <span class="st">"your@email.edu.au"</span>,</span>
<span>           atlas <span class="op">=</span> <span class="st">"ALA"</span>,</span>
<span>           ALA_taxon <span class="op">=</span> <span class="st">"Apiformes"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="import-and-merge-ala-scan-idigbio-and-gbif-data">1.2 Import and merge ALA, SCAN, iDigBio, and GBIF data<a class="anchor" aria-label="anchor" href="#import-and-merge-ala-scan-idigbio-and-gbif-data"></a>
</h3>
<p>If you are planning on combining data from ALA, SCAN, iDigBio, and/or
GBIF, <code><a href="../reference/repoMerge.html">BeeBDC::repoMerge()</a></code> is a handy function that should
help you with this and help you to extract all of the metadata and
citations that you need. Remember, that the main workflow only needs a
nice Darwin Core-formatted dataset with which to work!</p>
<p>Supply the path to where the data are, the save_type is either
“csv_files” or “R_file”.</p>
<pre><code><span>  <span class="va">DataImp</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/repoMerge.html">repoMerge</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>, </span>
<span>                  occ_paths <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/repoFinder.html">repoFinder</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span><span class="op">)</span>,</span>
<span>                  save_type <span class="op">=</span> <span class="st">"R_file"</span><span class="op">)</span></span></code></pre>
<p>If there is an error in finding a file, run <code><a href="../reference/repoFinder.html">repoFinder()</a></code>
by itself to troubleshoot. For example:</p>
<pre><code><span>            <span class="co">#BeeBDC::repoFinder(path = DataPath)</span></span>
<span>            <span class="co">#OUTPUT:</span></span>
<span>            <span class="co">#$ALA_data</span></span>
<span>            <span class="co">#[1] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/ALA_galah_path/galah_download_2022-09-15/data.csv"</span></span>
<span>  </span>
<span>            <span class="co">#$GBIF_data</span></span>
<span>            <span class="co">#[1] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0000165-220831081235567/occurrence.txt"</span></span>
<span>            <span class="co">#[2] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436695-210914110416597/occurrence.txt"</span></span>
<span>            <span class="co">#[3] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436697-210914110416597/occurrence.txt"</span></span>
<span>            <span class="co">#[4] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436704-210914110416597/occurrence.txt"</span></span>
<span>            <span class="co">#[5] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436732-210914110416597/occurrence.txt"</span></span>
<span>            <span class="co">#[6] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436733-210914110416597/occurrence.txt"</span></span>
<span>            <span class="co">#[7] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436734-210914110416597/occurrence.txt"</span></span>
<span>                    </span>
<span>            <span class="co">#$iDigBio_data</span></span>
<span>            <span class="co">#[1] "F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/iDigBio_webDL_30Aug2022/5aa5abe1-62e0-4d8c-bebf-4ac13bd9e56f/occurrence_raw.csv"</span></span>
<span>  </span>
<span>            <span class="co">#$SCAN_data</span></span>
<span>            <span class="co">#character(0)</span></span>
<span>            <span class="co">#Failing because SCAN_data seems to be missing. Downloaded separatly from the one drive</span></span></code></pre>
<p>Load in the most-recent version of these data if needed. This will
return a list with:</p>
<ol style="list-style-type: decimal">
<li><p>The occurrence dataset with attributes (.$Data_WebDL)</p></li>
<li>
<p>The appended eml file (.$eml_files)</p>
<pre><code><span><span class="va">DataImp</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/importOccurrences.html">importOccurrences</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>,</span>
<span>                       fileName <span class="op">=</span> <span class="st">"BeeData_"</span><span class="op">)</span></span></code></pre>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="save-data">1.5 Save data<a class="anchor" aria-label="anchor" href="#save-data"></a>
</h3>
<p>Choose the type of data format you want to use in saving your work in
1.x.</p>
<pre><code><span>  <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dataSaver.html">dataSaver</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>,<span class="co"># The main path to look for data in</span></span>
<span>       save_type <span class="op">=</span> <span class="st">"CSV_file"</span>, <span class="co"># "R_file" OR "CSV_file"</span></span>
<span>       occurrences <span class="op">=</span> <span class="va">Complete_data</span><span class="op">$</span><span class="va">Data_WebDL</span>, <span class="co"># The existing datasheet</span></span>
<span>       eml_files <span class="op">=</span> <span class="va">Complete_data</span><span class="op">$</span><span class="va">eml_files</span>, <span class="co"># The existing EML files</span></span>
<span>       file_prefix <span class="op">=</span> <span class="st">"Fin_"</span><span class="op">)</span> <span class="co"># The prefix for the fileNames</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">Complete_data</span>, <span class="va">DataImp</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="data-preparation">2.0 Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>The data preparation section of the script relates mostly to
integrating <strong>bee</strong> occurrence datasets and corrections and
so may be skipped by many general taxon users.</p>
<div class="section level3">
<h3 id="standardise-datasets">2.1 Standardise datasets<a class="anchor" aria-label="anchor" href="#standardise-datasets"></a>
</h3>
<p>You may either use:</p>
<ul>
<li><ol style="list-style-type: lower-alpha">
<li>the bdc import method (works well with general datasets)
<strong><em>or</em></strong>
</li>
</ol></li>
<li><ol start="2" style="list-style-type: lower-alpha">
<li>the jbd import method (works well with above data merge)</li>
</ol></li>
</ul>
<div class="section level4">
<h4 id="a--bdc-import">a. bdc import<a class="anchor" aria-label="anchor" href="#a--bdc-import"></a>
</h4>
<p>The bdc import is <strong>NOT</strong> truly supported here, but
provided as an example. Please go to section 2.1b below. Read in the
<strong>bdc</strong> metadata and standardise the dataset to bdc.</p>
<pre><code><span>        <span class="va">bdc_metadata</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">DataPath</span>, <span class="st">"out_file"</span>, <span class="st">"bdc_integration.csv"</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="co"># ?issue — datasetName is a darwinCore field already!</span></span>
<span>        <span class="co"># Standardise the dataset to bdc</span></span>
<span>        <span class="va">db_standardized</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_standardize_datasets.html" class="external-link">bdc_standardize_datasets</a></span><span class="op">(</span></span>
<span>          metadata <span class="op">=</span> <span class="va">bdc_metadata</span>,</span>
<span>          format <span class="op">=</span> <span class="st">"csv"</span>,</span>
<span>          overwrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          save_database <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="co"># read in configuration description file of the column header info</span></span>
<span>        <span class="va">config_description</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">DataPath</span>, <span class="st">"Output"</span>, <span class="st">"bdc_configDesc.csv"</span>,</span>
<span>                                                    sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, </span>
<span>                                              show_col_types <span class="op">=</span> <span class="cn">FALSE</span>, trim_ws <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  </span></code></pre>
</div>
<div class="section level4">
<h4 id="b--jbd-import">b. jbd import<a class="anchor" aria-label="anchor" href="#b--jbd-import"></a>
</h4>
<p>Find the path, read in the file, and add the <em>database_id</em>
column.</p>
<pre><code><span>  <span class="va">occPath</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/fileFinder.html">fileFinder</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>, fileName <span class="op">=</span> <span class="st">"Fin_BeeData_combined_"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span>  <span class="va">db_standardized</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">occPath</span>, </span>
<span>                                       <span class="co"># Use the basic ColTypeR function to determine types</span></span>
<span>                                     col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span>, trim_ws <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                                     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>database_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Dorey_data_"</span>, </span>
<span>                                     <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>                                     .before <span class="op">=</span> <span class="va">family</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="match-database_id">2.2 Match database_id<a class="anchor" aria-label="anchor" href="#match-database_id"></a>
</h4>
<p>If you have prior runs from which you’d like to match
<em>database_id</em>s with from the current run, you may use the below
script to try to match <em>database_id</em>s with prior runs.</p>
<p>Read in a prior run of choice.</p>
<pre><code><span>  <span class="va">priorRun</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/fileFinder.html">fileFinder</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>,</span>
<span>                          file <span class="op">=</span> <span class="st">"01_prefilter_database_9Aug22.csv"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">.</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>This function will attempt to find the <em>database_id</em>s from
prior runs.</p>
<pre><code><span>  <span class="va">db_standardized</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/idMatchR.html">idMatchR</a></span><span class="op">(</span></span>
<span>  currentData <span class="op">=</span> <span class="va">db_standardized</span>,</span>
<span>  priorData <span class="op">=</span> <span class="va">priorRun</span>,</span>
<span>    <span class="co"># First matches will be given preference over later ones</span></span>
<span>  matchBy <span class="op">=</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/lst.html" class="external-link">lst</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gbifID"</span>, <span class="st">"dataSource"</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"catalogNumber"</span>, <span class="st">"institutionCode"</span>, <span class="st">"dataSource"</span>, <span class="st">"decimalLatitude"</span>,</span>
<span>                          <span class="st">"decimalLongitude"</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"occurrenceID"</span>, <span class="st">"dataSource"</span>,<span class="st">"decimalLatitude"</span>,<span class="st">"decimalLongitude"</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"recordId"</span>, <span class="st">"dataSource"</span>,<span class="st">"decimalLatitude"</span>,<span class="st">"decimalLongitude"</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"dataSource"</span>,<span class="st">"decimalLatitude"</span>,<span class="st">"decimalLongitude"</span><span class="op">)</span>,</span>
<span>                        <span class="co"># Because INHS was entered as it's own dataset but is now included in the GBIF    download...</span></span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"catalogNumber"</span>, <span class="st">"institutionCode"</span>, <span class="st">"dataSource"</span>,</span>
<span>                          <span class="st">"decimalLatitude"</span>,<span class="st">"decimalLongitude"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="co"># You can exclude datasets from prior by matching their prefixs — before first underscore:</span></span>
<span>  excludeDataset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ASP"</span>, <span class="st">"BMin"</span>, <span class="st">"BMont"</span>, <span class="st">"CAES"</span>, <span class="st">"EaCO"</span>, <span class="st">"Ecd"</span>, <span class="st">"EcoS"</span>,</span>
<span>                     <span class="st">"Gai"</span>, <span class="st">"KP"</span>, <span class="st">"EPEL"</span>, <span class="st">"CAES"</span>, <span class="st">"EaCO"</span>, <span class="st">"FSCA"</span>, <span class="st">"SMC"</span>, <span class="st">"Lic"</span>, <span class="st">"Arm"</span>,</span>
<span>                     <span class="st">"VicWam"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span> <span class="co"># Remove redundant files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">priorRun</span><span class="op">)</span></span></code></pre>
<p>Save the dataset.</p>
<pre><code><span>  <span class="va">db_standardized</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>,</span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"00_prefilter_database.csv"</span>,</span>
<span>                           sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="initial-flags">3.0 Initial flags<a class="anchor" aria-label="anchor" href="#initial-flags"></a>
</h2>
<p>Now that we have some kind of dataset put together in Darwin Core
format, we can start to think about actually doing some filtering! Keep
in mind, as you progress through this vignette, that not all filters
might be relevant for your hypothesis! So, thinking about then ones that
you apply at the end throughout is probably a useful endeavor. However,
do keep in mind that the only thing that it costs you to run every
filter is time.</p>
<p>Read data back in if needed. OutPath_Intermediate (and a few other
directories) should be have been created and saved to the global
environment by <code><a href="../reference/dirMaker.html">dirMaker()</a></code>.</p>
<pre><code><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"db_standardized"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">db_standardized</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"00_prefilter_database.csv"</span>,</span>
<span>                                    sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">}</span></span></code></pre>
<p>Normally, you would use the full dataset, as read in above. But, for
the sake of this vignette, we will use a combination of two example
datasets. These example datasets can further be very useful for testing
functions if you’re ever feeling a bit confused and overwhelmed!</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bees3sp"</span>, package <span class="op">=</span> <span class="st">"BeeBDC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"beesRaw"</span>, package <span class="op">=</span> <span class="st">"BeeBDC"</span><span class="op">)</span></span>
<span><span class="va">db_standardized</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">beesRaw</span>, </span>
<span>                                      <span class="co"># Only keep a subset of columns from bees3sp</span></span>
<span>                             <span class="va">bees3sp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">beesRaw</span><span class="op">)</span><span class="op">)</span>, <span class="va">countryCode</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><em>For more details about the <strong>bdc</strong> package, please
see their <a href="https://brunobrr.github.io/bdc/articles/prefilter.html" class="external-link">tutorial.</a></em></p>
<div class="section level3">
<h3 id="sciname">3.1 SciName<a class="anchor" aria-label="anchor" href="#sciname"></a>
</h3>
<p>Flag occurrences with NO <em>scientificName</em> provided.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_scientificName_empty.html" class="external-link">bdc_scientificName_empty</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">db_standardized</span>, sci_name <span class="op">=</span> <span class="st">"scientificName"</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_scientificName_empty:</span></span>
<span><span class="co">## Flagged 0 records.</span></span>
<span><span class="co">## One column was added to the database.</span></span>
<span><span class="co"># now that this is saved, remove it to save space in memory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">db_standardized</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="misscoords">3.2 MissCoords<a class="anchor" aria-label="anchor" href="#misscoords"></a>
</h3>
<p>Flag occurrences with missing <em>decimalLatitude</em> and
<em>decimalLongitude</em>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_coordinates_empty.html" class="external-link">bdc_coordinates_empty</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>    lon <span class="op">=</span> <span class="st">"decimalLongitude"</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_coordinates_empty:</span></span>
<span><span class="co">## Flagged 42 records.</span></span>
<span><span class="co">## One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="outofrange">3.3 OutOfRange<a class="anchor" aria-label="anchor" href="#outofrange"></a>
</h3>
<p>Flag occurrences that are not on Earth (outside of -180 to 180 or -90
to 90 degrees).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_coordinates_outOfRange.html" class="external-link">bdc_coordinates_outOfRange</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>    lon <span class="op">=</span> <span class="st">"decimalLongitude"</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_coordinates_outOfRange:</span></span>
<span><span class="co">## Flagged 0 records.</span></span>
<span><span class="co">## One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="source">3.4 Source<a class="anchor" aria-label="anchor" href="#source"></a>
</h3>
<p>Flag occurrences that don’t match the <em>basisOfRecord</em> types
below.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_basisOfRecords_notStandard.html" class="external-link">bdc_basisOfRecords_notStandard</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_pf</span>,</span>
<span>  basisOfRecord <span class="op">=</span> <span class="st">"basisOfRecord"</span>,</span>
<span>  names_to_keep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="co"># Keep all plus some at the bottom.</span></span>
<span>    <span class="st">"Event"</span>,</span>
<span>    <span class="st">"HUMAN_OBSERVATION"</span>,</span>
<span>    <span class="st">"HumanObservation"</span>,</span>
<span>    <span class="st">"LIVING_SPECIMEN"</span>,</span>
<span>    <span class="st">"LivingSpecimen"</span>,</span>
<span>    <span class="st">"MACHINE_OBSERVATION"</span>,</span>
<span>    <span class="st">"MachineObservation"</span>,</span>
<span>    <span class="st">"MATERIAL_SAMPLE"</span>,</span>
<span>    <span class="st">"O"</span>,</span>
<span>    <span class="st">"Occurrence"</span>,</span>
<span>    <span class="st">"MaterialSample"</span>,</span>
<span>    <span class="st">"OBSERVATION"</span>,</span>
<span>    <span class="st">"Preserved Specimen"</span>,</span>
<span>    <span class="st">"PRESERVED_SPECIMEN"</span>,</span>
<span>    <span class="st">"preservedspecimen Specimen"</span>,</span>
<span>    <span class="st">"Preservedspecimen"</span>,</span>
<span>    <span class="st">"PreservedSpecimen"</span>,</span>
<span>    <span class="st">"preservedspecimen"</span>,</span>
<span>    <span class="st">"S"</span>,</span>
<span>    <span class="st">"Specimen"</span>,</span>
<span>    <span class="st">"Taxon"</span>,</span>
<span>    <span class="st">"UNKNOWN"</span>,</span>
<span>    <span class="st">""</span>,</span>
<span>    <span class="cn">NA</span>,</span>
<span>    <span class="st">"NA"</span>,</span>
<span>    <span class="st">"LITERATURE"</span>, </span>
<span>    <span class="st">"None"</span>, <span class="st">"Pinned Specimen"</span>, <span class="st">"Voucher reared"</span>, <span class="st">"Emerged specimen"</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_basisOfRecords_notStandard:</span></span>
<span><span class="co">## Flagged 1 of the following specific nature:</span></span>
<span><span class="co">##  MATERIAL_CITATION </span></span>
<span><span class="co">## One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="countryname">3.5 CountryName<a class="anchor" aria-label="anchor" href="#countryname"></a>
</h3>
<p>Try to harmonise country names.</p>
<div class="section level4">
<h4 id="a--prepare-dataset">a. prepare dataset<a class="anchor" aria-label="anchor" href="#a--prepare-dataset"></a>
</h4>
<p>Fix up country names based on common problems above and extract ISO2
codes for occurrences. The below are problems in the global bee dataset,
but you may find others in your own dataset. No harm in running the
below code and maybe altering it for your own use. Otherwise, don’t get
too caught up in it :)</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf_noNa</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/countryNameCleanR.html">countryNameCleanR</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_pf</span>,</span>
<span>    <span class="co"># Create a Tibble of common issues in country names and their replacements</span></span>
<span>  commonProblems <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>problem <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'U.S.A.'</span>, <span class="st">'US'</span>,<span class="st">'USA'</span>,<span class="st">'usa'</span>,<span class="st">'UNITED STATES'</span>,</span>
<span>                                              <span class="st">'United States'</span>,<span class="st">'U.S.A'</span>,<span class="st">'MX'</span>,<span class="st">'CA'</span>,<span class="st">'Bras.'</span>,<span class="st">'Braz.'</span>,</span>
<span>                                              <span class="st">'Brasil'</span>,<span class="st">'CNMI'</span>,<span class="st">'USA TERRITORY: PUERTO RICO'</span><span class="op">)</span>,</span>
<span>                                  fix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'United States of America'</span>,<span class="st">'United States of America'</span>,</span>
<span>                                          <span class="st">'United States of America'</span>,<span class="st">'United States of America'</span>,</span>
<span>                                          <span class="st">'United States of America'</span>,<span class="st">'United States of America'</span>,</span>
<span>                                          <span class="st">'United States of America'</span>,<span class="st">'Mexico'</span>,<span class="st">'Canada'</span>,<span class="st">'Brazil'</span>,</span>
<span>                                          <span class="st">'Brazil'</span>,<span class="st">'Brazil'</span>,<span class="st">'Northern Mariana Islands'</span>,<span class="st">'PUERTO.RICO'</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">##  - Using default country names and codes from https:en.wikipedia.org/wiki/ISO_3166-1_alpha-2 - static version from July 2022.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="b--run-function">b. run function<a class="anchor" aria-label="anchor" href="#b--run-function"></a>
</h4>
<p>Get country name from coordinates using a wrapper around the
<code>BeeBDC::jbd_country_from_coordinates()</code> function. Because
our dataset is much larger than those used to design
<strong>bdc</strong>, we have made it so that you can analyse data in
smaller pieces. Additionally, like some other functions in
<strong>BeeBDC</strong>, we have implemented parallel operations (using
mc.cores = #cores in stepSize = #rowsPerOperation); see
‘?<code><a href="../reference/jbd_CfC_chunker.html">BeeBDC::jbd_CfC_chunker()</a></code>’ for details. NOTE: In an
actual run you should use scale = “large”</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>  <span class="va">countryOutput</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/jbd_CfC_chunker.html">jbd_CfC_chunker</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf_noNa</span>,</span>
<span>                                   lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>                                   lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>                                   country <span class="op">=</span> <span class="st">"country"</span>,</span>
<span>                                    <span class="co"># How many rows to process at a time</span></span>
<span>                                   stepSize <span class="op">=</span> <span class="fl">1000000</span>,</span>
<span>                                    <span class="co"># Start row</span></span>
<span>                                   chunkStart <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                   path <span class="op">=</span> <span class="va">OutPath_Intermediate</span>,</span>
<span>                                    <span class="co"># Normally, please use scale = "large"</span></span>
<span>                                   scale <span class="op">=</span> <span class="st">"medium"</span>,</span>
<span>                                   mc.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  classes <span class="op">=</span> <span class="st">"warning"</span><span class="op">)</span></span>
<span><span class="co">##  - Starting parallel operation. Unlike the serial operation (mc.cores = 1) , a parallel operation will not provide running feedback. Please be patient  as this function may take some time to complete. Each chunk will be run on  a seperate thread so also be aware of RAM usage.</span></span>
<span><span class="co">##  - We have updated the country names of 39 occurrences that previously had no country name assigned.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="c--re-merge">c. re-merge<a class="anchor" aria-label="anchor" href="#c--re-merge"></a>
</h4>
<p>Join these datasets.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">check_pf</span>, <span class="va">countryOutput</span>, by <span class="op">=</span> <span class="st">"database_id"</span>, suffix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span>,</span>
<span>    <span class="st">"CO"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Take the new country name if the original is NA</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>country <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">country</span><span class="op">)</span>, <span class="va">countryCO</span>, <span class="va">country</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Remove duplicates if they arose from left_join!</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Save the dataset.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"01_prefilter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Re-load in if needed.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_pf"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">DataPath</span>, <span class="st">"Output"</span>, <span class="st">"Intermediate"</span>, <span class="st">"01_prefilter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Remove these interim datasets.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">check_pf_noNa</span>, <span class="va">countryOutput</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="standardconames">3.6 StandardCoNames<a class="anchor" aria-label="anchor" href="#standardconames"></a>
</h3>
<p>Run the function, which standardises country names and adds ISO2
codes, if needed.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Standardise country names and add ISO2 codes if needed</span></span>
<span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_country_standardized.html" class="external-link">bdc_country_standardized</a></span><span class="op">(</span></span>
<span>  <span class="co"># Remove the countryCode and country_suggested columns to avoid an error with </span></span>
<span>    <span class="co"># where two "countryCode" and "country_suggested" columns exist (i.e. if the dataset has been  </span></span>
<span>    <span class="co"># run before)</span></span>
<span>  data <span class="op">=</span> <span class="va">check_pf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">any_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"countryCode"</span>, <span class="st">"country_suggested"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  country <span class="op">=</span> <span class="st">"country"</span></span>
<span><span class="op">)</span> </span>
<span><span class="co">## Loading auxiliary data: country names</span></span>
<span><span class="co">## Standardizing country names</span></span>
<span><span class="co">## country found: Argentina</span></span>
<span><span class="co">## country found: Australia</span></span>
<span><span class="co">## country found: Belgium</span></span>
<span><span class="co">## country found: Brazil</span></span>
<span><span class="co">## country found: Canada</span></span>
<span><span class="co">## country found: Colombia</span></span>
<span><span class="co">## country found: Costa Rica</span></span>
<span><span class="co">## country found: Ecuador</span></span>
<span><span class="co">## country found: Estonia</span></span>
<span><span class="co">## country found: Finland</span></span>
<span><span class="co">## country found: France</span></span>
<span><span class="co">## country found: Germany</span></span>
<span><span class="co">## country found: Ireland</span></span>
<span><span class="co">## country found: Mexico</span></span>
<span><span class="co">## country found: Norway</span></span>
<span><span class="co">## country found: South Africa</span></span>
<span><span class="co">## country found: Sweden</span></span>
<span><span class="co">## country found: Switzerland</span></span>
<span><span class="co">## country found: United Kingdom</span></span>
<span><span class="co">## country found: United States of America</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_country_standardized:</span></span>
<span><span class="co">## The country names of 141 records were standardized.</span></span>
<span><span class="co">## Two columns ('country_suggested' and 'countryCode') were added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="transpcoords">3.7 TranspCoords<a class="anchor" aria-label="anchor" href="#transpcoords"></a>
</h3>
<p>Flag and correct records when <em>decimalLatitude</em> and
<em>decimalLongitude</em> appear to be transposed. We created this
chunked version of <code><a href="https://brunobrr.github.io/bdc/reference/bdc_coordinates_transposed.html" class="external-link">bdc::bdc_coordinates_transposed()</a></code>
because it is very RAM-heavy using our large bee dataset. Like many of
our other ‘jbd_…’ functions there are other improvements - e.g.,
parallel running.</p>
<p>NOTE: Usually you would use scale = “large”, which requires
rnaturalearthhires</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/jbd_Ctrans_chunker.html">jbd_Ctrans_chunker</a></span><span class="op">(</span></span>
<span>  <span class="co"># bdc_coordinates_transposed inputs</span></span>
<span>  data <span class="op">=</span> <span class="va">check_pf</span>,</span>
<span>  id <span class="op">=</span> <span class="st">"database_id"</span>,</span>
<span>  lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>  lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>  country <span class="op">=</span> <span class="st">"country"</span>,</span>
<span>  countryCode <span class="op">=</span> <span class="st">"countryCode"</span>,</span>
<span>  border_buffer <span class="op">=</span> <span class="fl">0.2</span>, <span class="co"># in decimal degrees (~22 km at the equator)</span></span>
<span>  save_outputs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sci_names <span class="op">=</span> <span class="st">"scientificName"</span>,</span>
<span>  <span class="co"># chunker inputs</span></span>
<span>  stepSize <span class="op">=</span> <span class="fl">1000000</span>,  <span class="co"># How many rows to process at a time</span></span>
<span>  chunkStart <span class="op">=</span> <span class="fl">1</span>,  <span class="co"># Start row</span></span>
<span>  append <span class="op">=</span> <span class="cn">FALSE</span>,  <span class="co"># If FALSE it may overwrite existing dataset</span></span>
<span>  progressiveSave <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    <span class="co"># In a normal run, please use scale = "large"</span></span>
<span>  scale <span class="op">=</span> <span class="st">"medium"</span>,</span>
<span>  path <span class="op">=</span> <span class="va">OutPath_Check</span>,</span>
<span>  mc.cores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span> </span>
<span><span class="co">##  - Running chunker with:</span></span>
<span><span class="co">## stepSize = 1,000,000</span></span>
<span><span class="co">## chunkStart = 1</span></span>
<span><span class="co">## chunkEnd = 1,000,000</span></span>
<span><span class="co">## append = FALSE</span></span>
<span><span class="co">##  - Starting chunk 1...</span></span>
<span><span class="co">## From 1 to 1,000,000</span></span>
<span><span class="co">##  - Finished chunk 1 of 1. Total records examined: 205</span></span></code></pre></div>
<p>Get a quick summary of the number of transposed records.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">check_pf</span><span class="op">$</span><span class="va">coordinates_transposed</span>, useNA <span class="op">=</span> <span class="st">"always"</span><span class="op">)</span></span></code></pre></div>
<p>Save the dataset.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"01_prefilter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Read the data in again if needed.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_pf"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"01_prefilter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="coord-country">3.8 Coord-country<a class="anchor" aria-label="anchor" href="#coord-country"></a>
</h3>
<p>Collect all country names in the <em>country_suggested</em> column.
We rebuilt a <strong>bdc</strong> function to flag occurrences where the
coordinates are inconsistent with the provided country name.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/jbd_coordCountryInconsistent.html">jbd_coordCountryInconsistent</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>    lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>, scale <span class="op">=</span> <span class="fl">50</span>, pointBuffer <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">##  - Downloading naturalearth map...</span></span>
<span><span class="co">## Spherical geometry (s2) switched off</span></span>
<span><span class="co">##  - Extracting initial country names without buffer...</span></span>
<span><span class="co">##  - Buffering naturalearth map by pointBuffer...</span></span>
<span><span class="co">## dist is assumed to be in decimal degrees (arc_degrees).</span></span>
<span><span class="co">##  - Extracting FAILED country names WITH buffer...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## jbd_coordinates_country_inconsistent:</span></span>
<span><span class="co">## Flagged 2 records.</span></span>
<span><span class="co">## The column, '.coordinates_country_inconsistent', was added to the database.</span></span>
<span><span class="co">##  - Completed in 0.8 secs</span></span></code></pre></div>
<p>Save the dataset.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"01_prefilter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="georefissue">3.9 GeoRefIssue<a class="anchor" aria-label="anchor" href="#georefissue"></a>
</h3>
<p>This function identifies records whose coordinates can potentially be
extracted from locality information, which must be manually checked
later.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xyFromLocality</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_coordinates_from_locality.html" class="external-link">bdc_coordinates_from_locality</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, locality <span class="op">=</span> <span class="st">"locality"</span>,</span>
<span>    lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>, lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>, save_outputs <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_coordinates_from_locality </span></span>
<span><span class="co">## Found 38 records missing or with invalid coordinates but with potentially useful information on locality.</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save the resultant data</span></span>
<span><span class="va">xyFromLocality</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Check</span>, <span class="st">"01_coordinates_from_locality.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Remove the spent data.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">xyFromLocality</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="flag-absent">3.10 Flag Absent<a class="anchor" aria-label="anchor" href="#flag-absent"></a>
</h3>
<p>Many of the large data repositories are happy to store presence and
absence records. This is very important to know, because if you leave
absence records in your dataset and analyse them as an occurrence
record… well, you’ll be doing a no-no to say the very least!</p>
<p>Flag the records marked as “absent”.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagAbsent.html">flagAbsent</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, PresAbs <span class="op">=</span> <span class="st">"occurrenceStatus"</span><span class="op">)</span></span>
<span><span class="co">## \.occurrenceAbsent:</span></span>
<span><span class="co">##  Flagged 8 absent records:</span></span>
<span><span class="co">##  One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="flag-license">3.11 flag License<a class="anchor" aria-label="anchor" href="#flag-license"></a>
</h3>
<p>Flag the records that may not be used according to their license
information.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagLicense.html">flagLicense</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>,</span>
<span>                    strings_to_restrict <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>                    <span class="co"># DON'T flag if in the following dataSource(s)</span></span>
<span>                    excludeDataSource <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">## \.unLicensed:</span></span>
<span><span class="co">##  Flagged 0 records that may NOT be used.</span></span>
<span><span class="co">##  One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gbif-issue">3.12 GBIF issue<a class="anchor" aria-label="anchor" href="#gbif-issue"></a>
</h3>
<p>Flag select issues that are flagged by GBIF.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/GBIFissues.html">GBIFissues</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, issueColumn <span class="op">=</span> <span class="st">"issue"</span>, GBIFflags <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"COORDINATE_INVALID"</span>,</span>
<span>    <span class="st">"ZERO_COORDINATE"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  - jbd_GBIFissues:</span></span>
<span><span class="co">## Flagged 0 </span></span>
<span><span class="co">##   The .GBIFflags column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="flag-reports">3.13 Flag Reports<a class="anchor" aria-label="anchor" href="#flag-reports"></a>
</h3>
<div class="section level4">
<h4 id="a--save-flags">a. Save flags<a class="anchor" aria-label="anchor" href="#a--save-flags"></a>
</h4>
<p>Save the flags so far. This function will make sure that you keep a
copy of everything that has been flagged up until now. This will be
updated throughout the script and can accessed at the end, so be wary of
moving files around manually. However, these data will also still be
maintained in the main running file, so this is an optional
fail-safe.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flagFile</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagRecorder.html">flagRecorder</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_pf</span>,</span>
<span>  outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Report</span>, sep <span class="op">=</span><span class="st">""</span><span class="op">)</span>,</span>
<span>  fileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"flagsRecorded_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>,  <span class="st">".csv"</span><span class="op">)</span>,</span>
<span>    <span class="co"># These are the columns that will be kept along with the flags</span></span>
<span>  idColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"database_id"</span>, <span class="st">"id"</span>, <span class="st">"catalogNumber"</span>, <span class="st">"occurrenceID"</span>, <span class="st">"dataSource"</span><span class="op">)</span>,</span>
<span>    <span class="co"># TRUE if you want to find a file from a previous part of the script to append to</span></span>
<span>  append <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Update the <em>.summary</em> column</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_pf</span>,</span>
<span>    <span class="co"># Don't filter these columns (or NULL)</span></span>
<span>  dontFilterThese <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    <span class="co"># Remove the filtering columns?</span></span>
<span>  removeFilterColumns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    <span class="co"># Filter to ONLY cleaned data?</span></span>
<span>  filterClean <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">##  - We will flag all columns starting with '.'</span></span>
<span><span class="co">##  - summaryFun:</span></span>
<span><span class="co">## Flagged 52 </span></span>
<span><span class="co">##   The .summary column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="c--reporting">c. Reporting<a class="anchor" aria-label="anchor" href="#c--reporting"></a>
</h4>
<p>Use <strong>bdc</strong> to generate reports.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">report</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_create_report.html" class="external-link">bdc_create_report</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_pf</span>, database_id <span class="op">=</span> <span class="st">"database_id"</span>, workflow_step <span class="op">=</span> <span class="st">"prefilter"</span>,</span>
<span>    save_report <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="save">3.14 Save<a class="anchor" aria-label="anchor" href="#save"></a>
</h3>
<p>Save the intermediate dataset.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_pf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"01_prefilter_output.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="taxonomy">4.0 Taxonomy<a class="anchor" aria-label="anchor" href="#taxonomy"></a>
</h2>
<p><em>For more information about the corresponding bdc functions used
in this section, see their <a href="https://brunobrr.github.io/bdc/articles/taxonomy.html" class="external-link">tutorial</a>.
</em></p>
<p>Read in the filtered dataset or rename the 3.x dataset for 4.0.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_pf"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">database</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"01_prefilter_output.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># OR rename and remove</span></span>
<span>    <span class="va">database</span> <span class="op">&lt;-</span> <span class="va">check_pf</span></span>
<span>    <span class="co"># Remove spent dataset</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">check_pf</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Remove <em>names_clean</em> if it already exists (i.e. you have run
the following functions before on this dataset before).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">database</span> <span class="op">&lt;-</span> <span class="va">database</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">any_of</a></span><span class="op">(</span><span class="st">"names_clean"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="prep-data-names">4.1 Prep data names<a class="anchor" aria-label="anchor" href="#prep-data-names"></a>
</h3>
<p>This step cleans the database’s <em>scientificName</em> column.</p>
<p><strong>! MAC</strong>: You might need to install
<strong>gnparser</strong> through terminal using <a href="https://brew.sh" class="external-link">homebrew</a> — brew brew tap gnames/gn brew
install gnparser</p>
<div class="alert alert-info">
<p><strong> Attention:</strong> <br> This can be difficult for a Windows
install. Ensure you have the most recent version of R, R Studio, and R
packages. Also, check package ‘<strong>rgnparser</strong>’ is installed
correctly. If you still can not get the below code to work, you may have
to download the latest version of ‘gnparser’ from <a href="https://github.com/gnames/gnparser/releases/tag/v1.6.9" class="external-link">here</a>.
You may then need to manually install it and edit your systems
environmental variable PATH to locate ‘gnparser.exe’. See <a href="https://github.com/gnames/gnparser#installation" class="external-link">here</a>.</p>
</div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parse_names</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_clean_names.html" class="external-link">bdc_clean_names</a></span><span class="op">(</span>sci_names <span class="op">=</span> <span class="va">database</span><span class="op">$</span><span class="va">scientificName</span>, save_outputs <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The latest gnparser version is v1.7.4</span></span>
<span><span class="co">## gnparser has been installed to /home/runner/bin</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## &gt;&gt; Family names prepended to scientific names were flagged and removed from 0 records.</span></span>
<span><span class="co">## &gt;&gt; Terms denoting taxonomic uncertainty were flagged and removed from 0 records.</span></span>
<span><span class="co">## &gt;&gt; Other issues, capitalizing the first letter of the generic name, replacing empty names by NA, and     removing extra spaces, were flagged and corrected or removed from 1 records.</span></span>
<span><span class="co">## &gt;&gt; Infraspecific terms were flagged and removed from 0 records.</span></span></code></pre>
<p>Keep only the <em>.uncer_terms</em> and <em>names_clean</em>
columns.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parse_names</span> <span class="op">&lt;-</span> <span class="va">parse_names</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.uncer_terms</span>, <span class="va">names_clean</span><span class="op">)</span></span></code></pre></div>
<p>Merge names with the complete dataset.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">database</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="va">database</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">parse_names</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="harmonise-taxonomy">4.2 Harmonise taxonomy<a class="anchor" aria-label="anchor" href="#harmonise-taxonomy"></a>
</h3>
<p>Download the custom taxonomy file from the <strong>BeeBDC</strong>
package and <a href="https://www.discoverlife.org" class="external-link">Discover Life</a>
website.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxonomyFile</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/beesTaxonomy.html">beesTaxonomy</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="alert alert-info">
<p><strong> Attention:</strong> <br> As of version 1.1.0,
<strong>BeeBDC</strong> now has a new function that can download
taxonomies using the <strong>taxadb</strong> package and transform them
into the <strong>BeeBDC</strong> format. The function,
<code><a href="../reference/taxadbToBeeBDC.html">BeeBDC::taxadbToBeeBDC()</a></code>, allows the user to choose their
desired provider (e.g., “gbif”, “itis”…), version, taxon name and rank,
and to save the taxonomy as a readable csv or not. For example for the
bee genus <em>Apis</em>:</p>
<pre><code><span><span class="va">ApisTaxonomy</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/taxadbToBeeBDC.html">taxadbToBeeBDC</a></span><span class="op">(</span></span>
<span>  name <span class="op">=</span> <span class="st">"Apis"</span>,</span>
<span>  rank <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span>  provider <span class="op">=</span> <span class="st">"gbif"</span>,</span>
<span>  version <span class="op">=</span> <span class="st">"22.12"</span>,</span>
<span>  outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  fileName <span class="op">=</span> <span class="st">"ApisTaxonomy.csv"</span></span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<p>Harmonise the names in the occurrence tibble. This flags the
occurrences without a matched name and matches names to their correct
name according to <a href="https://www.discoverlife.org" class="external-link">Discover
Life</a>. You can also use multiple cores to achieve this. See
‘?<code><a href="../reference/HarmoniseR.html">harmoniseR()</a></code>’ for details.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">database</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/HarmoniseR.html">harmoniseR</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>, <span class="co">#The path to a folder that the output can be saved</span></span>
<span>                       taxonomy <span class="op">=</span> <span class="va">taxonomyFile</span>, <span class="co"># The formatted taxonomy file</span></span>
<span>                       data <span class="op">=</span> <span class="va">database</span>,</span>
<span>                       mc.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">##  - Formatting taxonomy for matching...</span></span>
<span><span class="co">## The names_clean column was not found and will be temporarily copied from scientificName</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  - Harmonise the occurrence data with unambiguous names...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  - Attempting to harmonise the occurrence data with ambiguous names...</span></span>
<span><span class="co">##  - Formatting merged datasets...</span></span>
<span><span class="co">## Removing the names_clean column...</span></span>
<span><span class="co">##  - We matched valid names to 196 of 205 occurrence records. This leaves a total of 9 unmatched occurrence records.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## harmoniseR:</span></span>
<span><span class="co">## 9</span></span>
<span><span class="co">## records were flagged.</span></span>
<span><span class="co">## The column, '.invalidName' was added to the database.</span></span>
<span><span class="co">##  - We updated the following columns: scientificName, species, family, subfamily, genus, subgenus, specificEpithet, infraspecificEpithet, and scientificNameAuthorship. The previous scientificName column was converted to verbatimScientificName</span></span>
<span><span class="co">##  - Completed in 0.58 secs</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">taxonomyFile</span><span class="op">)</span></span></code></pre></div>
<p>Save the harmonised file.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">database</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">DataPath</span>, <span class="st">"Output"</span>, <span class="st">"Intermediate"</span>, <span class="st">"02_taxonomy_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="save-flags">4.3 Save flags<a class="anchor" aria-label="anchor" href="#save-flags"></a>
</h3>
<p>Save the flags so far. This will find the most-recent flag file and
append your new data to it. You can double-check the data and number of
columns if you’d like to be thorough and sure that all of data are
intact.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flagFile</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagRecorder.html">flagRecorder</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">database</span>, outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Report</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>, fileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"flagsRecorded_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">".csv"</span><span class="op">)</span>, idColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"database_id"</span>,</span>
<span>    <span class="st">"id"</span>, <span class="st">"catalogNumber"</span>, <span class="st">"occurrenceID"</span>, <span class="st">"dataSource"</span><span class="op">)</span>, append <span class="op">=</span> <span class="cn">TRUE</span>, printSummary <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="space">5.0 Space<a class="anchor" aria-label="anchor" href="#space"></a>
</h2>
<p><em>The final frontier or whatever.</em></p>
<p>Read in the latest database.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"database"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">database</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"02_taxonomy_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="coordinate-precision">5.1 Coordinate precision<a class="anchor" aria-label="anchor" href="#coordinate-precision"></a>
</h3>
<p>This function identifies records with a coordinate precision below a
specified number of decimal places. For example, the precision of a
coordinate with 1 decimal place is 11.132 km at the equator, i.e., the
scale of a large city. The major difference between the
<strong>bdc</strong> and <strong>BeeBDC</strong> functions is that
<code><a href="../reference/jbd_coordinates_precision.html">BeeBDC::jbd_coordinates_precision()</a></code> will only flag
occurrences if BOTH latitude and longitude are rounded (as opposed to
only one of these). This might be important where Excel has decided to
apply some rounding to one of either latitude or longitude…</p>
<p>Coordinates with one, two, or three decimal places present a
precision of ~11.1 km, ~1.1 km, and ~111 m at the equator,
respectively.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/jbd_coordinates_precision.html">jbd_coordinates_precision</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">database</span>, lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>    lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>, ndec <span class="op">=</span> <span class="fl">2</span>  <span class="co"># number of decimals to be tested</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## jbd_coordinates_precision:</span></span>
<span><span class="co">## Flagged 61 records</span></span>
<span><span class="co">## The '.rou' column was added to the database.</span></span></code></pre></div>
<p>Remove the spent dataset.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">database</span><span class="op">)</span></span></code></pre></div>
<p>Save the resulting file.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_inter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="common-spatial-issues">5.2 Common spatial issues<a class="anchor" aria-label="anchor" href="#common-spatial-issues"></a>
</h3>
<p>Only run for occurrences through <code>clean_coordinates()</code>
that are spatially ‘valid’.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tempSpace</span> <span class="op">&lt;-</span> <span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">.coordinates_empty</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">.coordinates_outOfRange</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Next, we will flag common spatial issues using functions of the
package <strong>CoordinateCleaner</strong>. It addresses some common
issues in biodiversity datasets.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tempSpace</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">CoordinateCleaner</span><span class="fu">::</span><span class="fu"><a href="https://ropensci.github.io/CoordinateCleaner/reference/clean_coordinates.html" class="external-link">clean_coordinates</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span>  <span class="va">tempSpace</span>,</span>
<span>    lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>    lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"scientificName"</span>,</span>
<span>    countries <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># Tests if coords are from x country. This is not needed.</span></span>
<span>    tests <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"capitals"</span>,     <span class="co"># records within 0.5 km of capitals centroids</span></span>
<span>      <span class="st">"centroids"</span>,    <span class="co"># records within 1 km around country and province centroids</span></span>
<span>      <span class="st">"equal"</span>,      <span class="co"># records with equal coordinates</span></span>
<span>      <span class="st">"gbif"</span>,         <span class="co"># records within 1 km of GBIF headquarters. (says 1 degree in package, but code says 1000 m)</span></span>
<span>      <span class="st">"institutions"</span>, <span class="co"># records within 100m of zoo and herbaria</span></span>
<span>      <span class="st">"zeros"</span>       <span class="co"># records with coordinates 0,0</span></span>
<span>      <span class="co"># "seas"        # Not flagged as this should be flagged by coordinate country inconsistent</span></span>
<span>    <span class="op">)</span>,</span>
<span>    capitals_rad <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    centroids_rad <span class="op">=</span> <span class="fl">500</span>,</span>
<span>    centroids_detail <span class="op">=</span> <span class="st">"both"</span>, <span class="co"># test both country and province centroids</span></span>
<span>    inst_rad <span class="op">=</span> <span class="fl">100</span>, <span class="co"># remove zoo and herbaria within 100m</span></span>
<span>    range_rad <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    zeros_rad <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    capitals_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    centroids_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    country_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    country_refcol <span class="op">=</span> <span class="st">"countryCode"</span>,</span>
<span>    inst_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    range_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    <span class="co"># seas_scale = 50,</span></span>
<span>    value <span class="op">=</span> <span class="st">"spatialvalid"</span> <span class="co"># result of tests are appended in separate columns</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="co"># Remove duplicate .summary column that can be replaced later and turn into a tibble</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">".summary"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Re-merge the datasets.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="va">tempSpace</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Re-bind with the records that were removed earlier</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">.coordinates_empty</span> <span class="op">==</span> <span class="cn">FALSE</span> <span class="op">|</span> <span class="va">.coordinates_outOfRange</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">tempSpace</span><span class="op">)</span></span></code></pre></div>
<p>Save the intermediate dataset.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_inter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="diagonal-grid">5.3 Diagonal + grid<a class="anchor" aria-label="anchor" href="#diagonal-grid"></a>
</h3>
<p>Finds sequential numbers that could be fill-down errors in lat and
long, and groups by the ‘groupingColumns’. This is accomplished by using
a sliding window with the length determined by minRepeats. Only
coordinates of precision ‘ndec’ (number of decimals in decimal degree
format) will be examined. Note, that this function is very RAM-intensive
and so the use of multiple threads should be approached with caution
depending on your dataset. However, the option is provided.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/diagonAlley.html">diagonAlley</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_space</span>,</span>
<span>  <span class="co"># The minimum number of repeats needed to find a sequence in for flagging</span></span>
<span>  minRepeats <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  ndec <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  groupingColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"eventDate"</span>, <span class="st">"recordedBy"</span>, <span class="st">"datasetName"</span><span class="op">)</span>,</span>
<span>  mc.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## Removing rounded coordinates with BeeBDC::jbd_coordinates_precision...</span></span>
<span><span class="co">## jbd_coordinates_precision:</span></span>
<span><span class="co">## Removed 68 records.</span></span>
<span><span class="co">##  - Merging results and adding the .sequential column...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## diagonAlley:</span></span>
<span><span class="co">## Flagged 0 records</span></span>
<span><span class="co">## The .sequential column was added to the database.</span></span>
<span><span class="co">##  - Completed in 0.03 secs</span></span></code></pre></div>
<p>Spatial gridding from rasterisation: Select only the records with
more than <code>X</code> occurrences.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">griddingDF</span> <span class="op">&lt;-</span> <span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Exclude NA lat and lon values</span></span>
<span><span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html" class="external-link">drop_na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"decimalLatitude"</span>, <span class="st">"decimalLongitude"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Group by the dataset name</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">datasetName</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Remove rows that aren't unique for lat and long</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">decimalLongitude</span>, <span class="va">decimalLatitude</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Find the groups with 4 or more occurrence records</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Run the gridding analysis to find datasets that might be gridded.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gridded_datasets</span> <span class="op">&lt;-</span> <span class="fu">CoordinateCleaner</span><span class="fu">::</span><span class="fu"><a href="https://ropensci.github.io/CoordinateCleaner/reference/cd_round.html" class="external-link">cd_round</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">griddingDF</span>, lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>    lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>, ds <span class="op">=</span> <span class="st">"datasetName"</span>, T1 <span class="op">=</span> <span class="fl">7</span>, min_unique_ds_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    test <span class="op">=</span> <span class="st">"both"</span>, value <span class="op">=</span> <span class="st">"dataset"</span>, graphs <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>, reg_out_thresh <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    reg_dist_min <span class="op">=</span> <span class="fl">0.1</span>, reg_dist_max <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># The griddingDF is no longer needed. remove it.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">griddingDF</span><span class="op">)</span></span></code></pre></div>
<p>Integrate these results with the main dataset.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Join the datasets</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>    <span class="co"># Select the columns of interest</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gridded_datasets</span>, <span class="va">dataset</span>, <span class="va">lon.flag</span>, <span class="va">lat.flag</span>, <span class="va">summary</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"datasetName"</span> <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Make new columns with more-consistent naming and change the NA vlaues to = TRUE (not flagged)</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.lonFlag <span class="op">=</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/replace_na.html" class="external-link">replace_na</a></span><span class="op">(</span><span class="va">lon.flag</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                .latFlag <span class="op">=</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/replace_na.html" class="external-link">replace_na</a></span><span class="op">(</span><span class="va">lat.flag</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                .gridSummary <span class="op">=</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/replace_na.html" class="external-link">replace_na</a></span><span class="op">(</span><span class="va">summary</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Remove old columns</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">lon.flag</span>, <span class="va">lat.flag</span>, <span class="va">summary</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Save the gridded_datasets file for later examination.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gridded_datasets</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_griddedDatasets.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">gridded_datasets</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="uncertainty">5.4 Uncertainty<a class="anchor" aria-label="anchor" href="#uncertainty"></a>
</h3>
<p>Flag records that exceed a <em>coordinateUncertaintyInMeters</em>
threshold in meters.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/coordUncerFlagR.html">coordUncerFlagR</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_space</span>, uncerColumn <span class="op">=</span> <span class="st">"coordinateUncertaintyInMeters"</span>,</span>
<span>    threshold <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">## \coordUncerFlagR:</span></span>
<span><span class="co">##  Flagged 33 geographically uncertain records:</span></span>
<span><span class="co">##  The column '.uncertaintyThreshold' was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="country-continent-checklists">5.5 Country &amp; continent checklists<a class="anchor" aria-label="anchor" href="#country-continent-checklists"></a>
</h3>
<p>This step identifies mismatches between the <a href="https://www.discoverlife.org" class="external-link">Discover Life</a> country checklist
— <code>beesChecklist</code> — for bee species and the dataset,
identifying potential misidentifications, outliers, etc.’</p>
<p>I would like to eventually add checklists for other taxa or get help
in this; if anyone is willing!</p>
<p>Download the <strong>bee</strong> country-level checklist.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">checklistFile</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/beesChecklist.html">beesChecklist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load in the small test dataset in the background</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"testChecklist.rda"</span>, package <span class="op">=</span> <span class="st">"BeeBDC"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Rename the file</span></span>
<span><span class="va">taxonomyFile</span> <span class="op">&lt;-</span> <span class="va">testChecklist</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">testChecklist</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/countryOutlieRs.html">countryOutlieRs</a></span><span class="op">(</span>checklist <span class="op">=</span> <span class="va">checklistFile</span>,</span>
<span>                        data <span class="op">=</span> <span class="va">check_space</span>,</span>
<span>                        keepAdjacentCountry <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        pointBuffer <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                          <span class="co"># Scale of map to return, one of 110, 50, 10 OR 'small', 'medium', 'large'</span></span>
<span>                          <span class="co"># Smaller numbers will result in much longer calculation times. </span></span>
<span>                          <span class="co"># We have not attempted a scale of 10.</span></span>
<span>                        scale <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                        mc.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## Error: object 'checklistFile' not found</span></span></code></pre></div>
<p>Since version 1.1.2 a new function,
<code><a href="../reference/continentOutlieRs.html">BeeBDC::continentOutlieRs()</a></code>, has been released which is
conceptually the same as <code><a href="../reference/countryOutlieRs.html">BeeBDC::countryOutlieRs()</a></code> except
it works on the level of continents. Users might appreciate that
generating a continent-level checklist is much easier than a
country-level one and so might actually be of more wide-spread use
(beyond bees).</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/continentOutlieRs.html">continentOutlieRs</a></span><span class="op">(</span>checklist <span class="op">=</span> <span class="va">checklistFile</span>,</span>
<span>                        data <span class="op">=</span> <span class="va">check_space</span>,</span>
<span>                        keepAdjacentContinent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        pointBuffer <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                          <span class="co"># Scale of map to return, one of 110, 50, 10 OR 'small', 'medium', 'large'</span></span>
<span>                          <span class="co"># Smaller numbers will result in much longer calculation times. </span></span>
<span>                          <span class="co"># We have not attempted a scale of 10.</span></span>
<span>                        scale <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                        mc.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## Error: object 'checklistFile' not found</span></span></code></pre></div>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A list of failed species-country combinations and their numbers can be output</span></span>
<span><span class="co"># here</span></span>
<span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">.countryOutlier</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">database_id</span>, <span class="va">scientificName</span>, <span class="va">country</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">scientificName</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>count_scientificName <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">scientificName</span>, <span class="va">country</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_failedCountryChecklist.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="map-spatial-errors">5.6 Map spatial errors<a class="anchor" aria-label="anchor" href="#map-spatial-errors"></a>
</h3>
<p>Assemble maps of potential spatial errors and outliers, either one
flag at a time or using the <em>.summary</em> column. First, you need to
rebuild the <em>.summary</em> column.</p>
<p>Rebuild the <em>.summary</em> column.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_space</span>, dontFilterThese <span class="op">=</span> <span class="cn">NULL</span>, removeFilterColumns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    filterClean <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Use col_to_map in order to map ONE spatial flag at a time or map the
<em>.summary</em> column for all flags.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">.summary</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># map only records flagged as FALSE</span></span>
<span>  <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_quickmap.html" class="external-link">bdc_quickmap</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">.</span>,</span>
<span>    lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>    lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>    col_to_map <span class="op">=</span> <span class="st">".summary"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.9</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="space-report">5.7 Space report<a class="anchor" aria-label="anchor" href="#space-report"></a>
</h3>
<p>Create the space report using <strong>bdc</strong>.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">report</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_create_report.html" class="external-link">bdc_create_report</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">.uncer_terms</span><span class="op">)</span><span class="op">)</span>, database_id <span class="op">=</span> <span class="st">"database_id"</span>, workflow_step <span class="op">=</span> <span class="st">"space"</span>,</span>
<span>    save_report <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="space-figures">5.8 Space figures<a class="anchor" aria-label="anchor" href="#space-figures"></a>
</h3>
<p>Create figures for the spatial data filtering results.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">figures</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/jbd_create_figures.html">jbd_create_figures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">.uncer_terms</span><span class="op">)</span><span class="op">)</span>, path <span class="op">=</span> <span class="va">DataPath</span>, database_id <span class="op">=</span> <span class="st">"database_id"</span>,</span>
<span>    workflow_step <span class="op">=</span> <span class="st">"space"</span>, save_figures <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For examining the figures, the options are:</p>
<ul>
<li><p><em>.cap</em> = Records around country capital centroid<br></p></li>
<li><p><em>.cen</em> = Records around country or province
centroids</p></li>
<li><p><em>.dbl</em> = Duplicated coordinates per species</p></li>
<li><p><em>.equ</em> = Identical coordinates</p></li>
<li><p><em>.otl</em> = Geographical outliers</p></li>
<li><p><em>.gbf</em> = Records around the GBIF headquarters</p></li>
<li><p><em>.inst</em> = Records around biodiversity
institutions</p></li>
<li><p><em>.rou</em> = Rounded (probably imprecise) coordinates</p></li>
<li>
<p><em>.urb</em> = Records within urban areas — (Likely not relevant
for bees.)</p>
<pre><code>      You can examine these figures, for example, by running:

      figures$.rou</code></pre>
</li>
</ul>
<p>Save interim dataset.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_inter_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="save-flags-1">5.9 Save flags<a class="anchor" aria-label="anchor" href="#save-flags-1"></a>
</h3>
<p>Save the flags so far.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagRecorder.html">flagRecorder</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_space</span>, outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Report</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>    fileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"flagsRecorded_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">".csv"</span><span class="op">)</span>, idColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"database_id"</span>,</span>
<span>        <span class="st">"id"</span>, <span class="st">"catalogNumber"</span>, <span class="st">"occurrenceID"</span>, <span class="st">"dataSource"</span><span class="op">)</span>, append <span class="op">=</span> <span class="cn">TRUE</span>, printSummary <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="save-1">5.10 Save<a class="anchor" aria-label="anchor" href="#save-1"></a>
</h3>
<p>Save the intermediate dataset.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_space</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="time">6.0 Time<a class="anchor" aria-label="anchor" href="#time"></a>
</h2>
<p>Read in the last database, if needed.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_space"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"03_space_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">check_time</span> <span class="op">&lt;-</span> <span class="va">check_space</span></span>
<span>    <span class="co"># Remove the spent file</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">check_space</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>You can plot a histogram of dates here.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu">lubridate</span><span class="fu">::</span><span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd_hms.html" class="external-link">ymd_hms</a></span><span class="op">(</span><span class="va">check_time</span><span class="op">$</span><span class="va">eventDate</span>, truncated <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, breaks <span class="op">=</span> <span class="fl">20</span>, main <span class="op">=</span> <span class="st">"Histogram of eventDates"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BeeBDC_main_files/figure-html/6.0ii-1.png" width="700"></p>
<p>Filter some silly dates that don’t make sense.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span><span class="op">$</span><span class="va">year</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">check_time</span><span class="op">$</span><span class="va">year</span> <span class="op">&gt;</span> <span class="fu">lubridate</span><span class="fu">::</span><span class="fu"><a href="https://lubridate.tidyverse.org/reference/year.html" class="external-link">year</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|</span> <span class="va">check_time</span><span class="op">$</span><span class="va">year</span> <span class="op">&lt;</span></span>
<span>    <span class="fl">1600</span>, <span class="cn">NA</span>, <span class="va">check_time</span><span class="op">$</span><span class="va">year</span><span class="op">)</span></span>
<span><span class="va">check_time</span><span class="op">$</span><span class="va">month</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">check_time</span><span class="op">$</span><span class="va">month</span> <span class="op">&gt;</span> <span class="fl">12</span> <span class="op">|</span> <span class="va">check_time</span><span class="op">$</span><span class="va">month</span> <span class="op">&lt;</span> <span class="fl">1</span>, <span class="cn">NA</span>, <span class="va">check_time</span><span class="op">$</span><span class="va">month</span><span class="op">)</span></span>
<span><span class="va">check_time</span><span class="op">$</span><span class="va">day</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">check_time</span><span class="op">$</span><span class="va">day</span> <span class="op">&gt;</span> <span class="fl">31</span> <span class="op">|</span> <span class="va">check_time</span><span class="op">$</span><span class="va">day</span> <span class="op">&lt;</span> <span class="fl">1</span>, <span class="cn">NA</span>, <span class="va">check_time</span><span class="op">$</span><span class="va">day</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="recover-dates">6.1 Recover dates<a class="anchor" aria-label="anchor" href="#recover-dates"></a>
</h3>
<p>The <code><a href="../reference/dateFindR.html">dateFindR()</a></code> function will search through some other
columns in order to find and rescue dates that may not have made it into
the correct columns. It will further update the <em>eventDate</em>,
<em>day</em>, <em>month</em>, and <em>year</em> columns where these data
were a) missing and b) located in one of the searched columns.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dateFindR.html">dateFindR</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>,</span>
<span>                        <span class="co"># Years above this are removed (from the recovered dates only)</span></span>
<span>                        maxYear <span class="op">=</span> <span class="fu">lubridate</span><span class="fu">::</span><span class="fu"><a href="https://lubridate.tidyverse.org/reference/year.html" class="external-link">year</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        <span class="co"># Years below this are removed (from the recovered dates only)</span></span>
<span>                        minYear <span class="op">=</span> <span class="fl">1700</span><span class="op">)</span></span>
<span><span class="co">##  - Preparing data...</span></span>
<span><span class="co">##  - Extracting dates from year, month, day columns...</span></span>
<span><span class="co">##  - Extracting dates from fieldNotes, locationRemarks, and verbatimEventDate columns in unambiguous ymd, dmy, mdy, and my formats...</span></span>
<span><span class="co">##  - Extracting year from fieldNotes, locationRemarks, and verbatimEventDate columns in ambiguous formats...</span></span>
<span><span class="co">##  - Formating and combining the new data..</span></span>
<span><span class="co">##  - Merging all data, nearly there...</span></span>
<span><span class="co">##  - Finished. </span></span>
<span><span class="co">## We now have 1 more full eventDate cells than in the input data.</span></span>
<span><span class="co">## We modified dates in </span></span>
<span><span class="co">## 174 occurrences.</span></span>
<span><span class="co">##  - As it stands, there are 174 complete eventDates and 31 missing dates.</span></span>
<span><span class="co">##  - There are also 175 complete year occurrences to filter from. This is up from an initial count of 174 At this rate, you will stand to lose 30 occurrences on the basis of missing year - Operation time: 0.595602035522461 secs</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="no-eventdate">6.2 No eventDate<a class="anchor" aria-label="anchor" href="#no-eventdate"></a>
</h3>
<p>Flag records that simply lack collection date. :(</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_eventDate_empty.html" class="external-link">bdc_eventDate_empty</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, eventDate <span class="op">=</span> <span class="st">"eventDate"</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_eventDate_empty:</span></span>
<span><span class="co">## Flagged 31 records.</span></span>
<span><span class="co">## One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="old-records">6.3 Old records<a class="anchor" aria-label="anchor" href="#old-records"></a>
</h3>
<p>This will flag records prior to the date selected. The year 1970 is
frequently chosen for species distribution modelling work. You may not
need to filter old records at all, so think critically about your use.
We have chosen 1950 as a lower extreme.</p>
<p>Many other works probably won’t need to apply a year filter at all;
for example, if you want to make a distribution map for a taxonomic
work.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_year_outOfRange.html" class="external-link">bdc_year_outOfRange</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, eventDate <span class="op">=</span> <span class="st">"year"</span>, year_threshold <span class="op">=</span> <span class="fl">1950</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## bdc_year_outOfRange:</span></span>
<span><span class="co">## Flagged 21 records.</span></span>
<span><span class="co">## One column was added to the database.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="time-report">6.4 Time report<a class="anchor" aria-label="anchor" href="#time-report"></a>
</h3>
<p><em>Not all of time, just the time pertaining to our precise
occurrence records.</em> Update the <em>.summary</em> column.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_time</span>,</span>
<span>  <span class="co"># Don't filter these columns (or NULL)</span></span>
<span>  dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>, <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span><span class="op">)</span>,</span>
<span>  <span class="co"># Remove the filtering columns?</span></span>
<span>  removeFilterColumns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="co"># Filter to ONLY cleaned data?</span></span>
<span>  filterClean <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">##  - We will NOT flag the following columns. However, they will remain in the data file.</span></span>
<span><span class="co">## .gridSummary, .lonFlag, .latFlag, .uncer_terms</span></span>
<span><span class="co">##  - summaryFun:</span></span>
<span><span class="co">## Flagged 116 </span></span>
<span><span class="co">##   The .summary column was added to the database.</span></span></code></pre></div>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">report</span> <span class="op">&lt;-</span> <span class="fu">bdc</span><span class="fu">::</span><span class="fu"><a href="https://brunobrr.github.io/bdc/reference/bdc_create_report.html" class="external-link">bdc_create_report</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, database_id <span class="op">=</span> <span class="st">"database_id"</span>,</span>
<span>    workflow_step <span class="op">=</span> <span class="st">"time"</span>, save_report <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="time-figures">6.5 Time figures<a class="anchor" aria-label="anchor" href="#time-figures"></a>
</h3>
<p>Create time results figures.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">figures</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/jbd_create_figures.html">jbd_create_figures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, path <span class="op">=</span> <span class="va">DataPath</span>, database_id <span class="op">=</span> <span class="st">"database_id"</span>,</span>
<span>    workflow_step <span class="op">=</span> <span class="st">"time"</span>, save_figures <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>You can check figures by using…</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">figures</span><span class="op">$</span><span class="va">year</span></span></code></pre></div>
<p>Save the time-revised data into the intermediate folder.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"04_time_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="save-flags-2">6.6 Save flags<a class="anchor" aria-label="anchor" href="#save-flags-2"></a>
</h3>
<p>Save the flags so far.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagRecorder.html">flagRecorder</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Report</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>    fileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"flagsRecorded_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">".csv"</span><span class="op">)</span>, idColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"database_id"</span>,</span>
<span>        <span class="st">"id"</span>, <span class="st">"catalogNumber"</span>, <span class="st">"occurrenceID"</span>, <span class="st">"dataSource"</span><span class="op">)</span>, append <span class="op">=</span> <span class="cn">TRUE</span>, printSummary <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="de-duplication">7.0 De-duplication<a class="anchor" aria-label="anchor" href="#de-duplication"></a>
</h2>
<p>The dataset can be re-read here if it does not already exist.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_time"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"04_time_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="deduplicate">7.1 deDuplicate<a class="anchor" aria-label="anchor" href="#deduplicate"></a>
</h3>
<p>We will FLAG duplicates here. These input columns can be hacked to
de-duplicate as you wish. This function uses user-specified inputs and
columns to identify duplicate occurrence records. Duplicates are
identified iteratively and will be tallied up, duplicate pairs
clustered, and sorted at the end of the function. The function is
designed to work with Darwin Core data with a <em>database_id</em>
column, but it is also modifiable to work with other columns. I would
encourage you to see ‘?<code><a href="../reference/dupeSummary.html">BeeBDC::dupeSummary()</a></code>’ for more
details as this function is quite modifiable to user needs.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dupeSummary.html">dupeSummary</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_time</span>,</span>
<span>  path <span class="op">=</span> <span class="va">OutPath_Report</span>,</span>
<span>   <span class="co"># options are "ID","collectionInfo", or "both"</span></span>
<span>  duplicatedBy <span class="op">=</span> <span class="st">"collectionInfo"</span>, </span>
<span>    <span class="co"># The columns to generate completeness info from (and to sort by completness)</span></span>
<span>  completeness_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"decimalLatitude"</span>,  <span class="st">"decimalLongitude"</span>,</span>
<span>                        <span class="st">"scientificName"</span>, <span class="st">"eventDate"</span><span class="op">)</span>,</span>
<span>   <span class="co"># The columns to ADDITIONALLY consider when finding duplicates in collectionInfo</span></span>
<span>  collectionCols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"decimalLatitude"</span>, <span class="st">"decimalLongitude"</span>, <span class="st">"scientificName"</span>, <span class="st">"eventDate"</span>, </span>
<span>                     <span class="st">"recordedBy"</span><span class="op">)</span>,</span>
<span>    <span class="co"># The columns to combine, one-by-one with the collectionCols</span></span>
<span>  collectInfoColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"catalogNumber"</span>, <span class="st">"otherCatalogNumbers"</span><span class="op">)</span>,</span>
<span>    <span class="co"># Custom comparisons — as a list of columns to compare</span></span>
<span>     <span class="co"># RAW custom comparisons do not use the character and number thresholds</span></span>
<span>  CustomComparisonsRAW <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/lst.html" class="external-link">lst</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"catalogNumber"</span>, <span class="st">"institutionCode"</span>, <span class="st">"scientificName"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>     <span class="co"># Other custom comparisons use the character and number thresholds</span></span>
<span>  CustomComparisons <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/lst.html" class="external-link">lst</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gbifID"</span>, <span class="st">"scientificName"</span><span class="op">)</span>,</span>
<span>                                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"occurrenceID"</span>, <span class="st">"scientificName"</span><span class="op">)</span>,</span>
<span>                                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"recordId"</span>, <span class="st">"scientificName"</span><span class="op">)</span>,</span>
<span>                                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"scientificName"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>   <span class="co"># The order in which you want to KEEP duplicated based on data source</span></span>
<span>   <span class="co"># try unique(check_time$dataSource)</span></span>
<span>  sourceOrder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GBIF"</span>,<span class="st">"SCAN"</span>,<span class="st">"iDigBio"</span><span class="op">)</span>,</span>
<span>    <span class="co"># Set the complexity threshold for id letter and number length</span></span>
<span>     <span class="co"># minimum number of characters when WITH the numberThreshold</span></span>
<span>  characterThreshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     <span class="co"># minimum number of numbers when WITH the characterThreshold</span></span>
<span>  numberThreshold <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     <span class="co"># Minimum number of numbers WITHOUT any characters</span></span>
<span>  numberOnlyThreshold <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># END dupeSummary</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  - Generating a basic completeness summary from the decimalLatitude, decimalLongitude, scientificName, eventDate columns.</span></span>
<span><span class="co">## This summary is simply the sum of complete.cases in each column. It ranges from zero to the N of columns. This will be used to sort duplicate rows and select the most-complete rows.</span></span>
<span><span class="co">##  - Updating the .summary column to sort by...</span></span>
<span><span class="co">##  - We will NOT flag the following columns. However, they will remain in the data file.</span></span>
<span><span class="co">## .gridSummary, .lonFlag, .latFlag, .uncer_terms, .uncertaintyThreshold, .unLicensed</span></span>
<span><span class="co">##  - summaryFun:</span></span>
<span><span class="co">## Flagged 93 </span></span>
<span><span class="co">##   The .summary column was added to the database.</span></span>
<span><span class="co">##  - Working on CustomComparisonsRAW duplicates...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 1 of 1:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the column(s): </span></span>
<span><span class="co">## catalogNumber, institutionCode, scientificName</span></span>
<span><span class="co">##  - Working on CustomComparisons duplicates...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 1 of 4:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the column(s): </span></span>
<span><span class="co">## gbifID, scientificName</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 2 of 4:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the column(s): </span></span>
<span><span class="co">## occurrenceID, scientificName</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 3 of 4:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the column(s): </span></span>
<span><span class="co">## recordId, scientificName</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 4 of 4:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the column(s): </span></span>
<span><span class="co">## id, scientificName</span></span>
<span><span class="co">##  - Working on collectionInfo duplicates...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 1 of 2:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the columns: </span></span>
<span><span class="co">## decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and catalogNumber</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Completed iteration 2 of 2:</span></span>
<span><span class="co">##  - Identified 0 duplicate records and kept 0 unique records using the columns: </span></span>
<span><span class="co">## decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and otherCatalogNumbers</span></span>
<span><span class="co">##  - Clustering duplicate pairs...</span></span>
<span><span class="co">## Duplicate pairs clustered. There are 0 duplicates across 0 kept duplicates.</span></span>
<span><span class="co">##  - Ordering data by 1. dataSource, 2. completeness and 3. .summary column...</span></span>
<span><span class="co">##  - Find and FIRST duplicate to keep and assign other associated duplicates to that one (i.e., across multiple tests a 'kept duplicate', could otherwise be removed)...</span></span>
<span><span class="co">##  - Duplicates have been saved in the file and location: /tmp/Rtmp4RrFst/Data_acquisition_workflow/Output/ReportduplicateRun_collectionInfo_2025-08-24.csv</span></span>
<span><span class="co">##  - Across the entire dataset, there are now 0 duplicates from a total of 205 occurrences.</span></span>
<span><span class="co">##  - Completed in 0.28 secs</span></span></code></pre></div>
<p>Save the dataset into the intermediate folder.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">check_time</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"04_2_dup_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="save-flags-3">7.2 Save flags<a class="anchor" aria-label="anchor" href="#save-flags-3"></a>
</h3>
<p>Save the flags so far.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagRecorder.html">flagRecorder</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Report</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>    fileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"flagsRecorded_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">".csv"</span><span class="op">)</span>, idColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"database_id"</span>,</span>
<span>        <span class="st">"id"</span>, <span class="st">"catalogNumber"</span>, <span class="st">"occurrenceID"</span>, <span class="st">"dataSource"</span><span class="op">)</span>, append <span class="op">=</span> <span class="cn">TRUE</span>, printSummary <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-filtering">8.0 Data filtering<a class="anchor" aria-label="anchor" href="#data-filtering"></a>
</h2>
<p>The dataset can be re-read here if it does not already exist.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_time"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"04_2_dup_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="rm-outliers">8.1 rm Outliers<a class="anchor" aria-label="anchor" href="#rm-outliers"></a>
</h3>
<p>Read in the most-recent duplicates file (generated by
<code><a href="../reference/dupeSummary.html">BeeBDC::dupeSummary()</a></code>) in order to identify the duplicates
of the expert outliers.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"duplicates"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">duplicates</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/fileFinder.html">fileFinder</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>, fileName <span class="op">=</span> <span class="st">"duplicateRun_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">##  - Dates found in file name(s). Finding most-recent file from file name...</span></span>
<span><span class="co">##  - Found the following file(s): </span></span>
<span><span class="co">##  /tmp/Rtmp4RrFst/Data_acquisition_workflow/Output/Report/duplicateRun_collectionInfo_2025-08-24.csv</span></span>
<span><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">0</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">19</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">## <span style="color: #BB0000;">chr</span> (19): group, database_id, database_id_keep, dupColumn_s, decimalLatitude...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<p>Identify the outliers and get a list of their database_ids. This
would require the source outlier files provided with the <a href="https://doi.org/10.1101/2023.06.30.547152" class="external-link">BeeBDC</a> paper. These
files can further be modified to include more outliers.</p>
<pre><code><span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/manualOutlierFindeR.html">manualOutlierFindeR</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_time</span>,</span>
<span>  DataPath <span class="op">=</span> <span class="va">DataPath</span>,</span>
<span>  PaigeOutliersName <span class="op">=</span> <span class="st">"removedBecauseDeterminedOutlier.csv"</span>,</span>
<span>  newOutliersName <span class="op">=</span> <span class="st">"^All_outliers_ANB_14March.xlsx"</span>,</span>
<span>  ColombiaOutliers_all <span class="op">=</span> <span class="st">"All_Colombian_OutlierIDs.csv"</span>,</span>
<span>  <span class="co"># A .csv with manual outlier records that are too close to otherwise TRUE records</span></span>
<span>  NearTRUE <span class="op">=</span> <span class="st">"nearTRUE.csv"</span>,</span>
<span>  duplicates <span class="op">=</span> <span class="va">duplicates</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="save-uncleaned">8.2 Save uncleaned<a class="anchor" aria-label="anchor" href="#save-uncleaned"></a>
</h3>
<p>Save the uncleaned dataset.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make sure that the .summary column is updated</span></span>
<span><span class="va">check_time</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">check_time</span>, dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>,</span>
<span>    <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span>, <span class="st">".uncertaintyThreshold"</span><span class="op">)</span>, removeFilterColumns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    filterClean <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">##  - We will NOT flag the following columns. However, they will remain in the data file.</span></span>
<span><span class="co">## .gridSummary, .lonFlag, .latFlag, .uncer_terms, .uncertaintyThreshold</span></span>
<span><span class="co">##  - summaryFun:</span></span>
<span><span class="co">## Flagged 93 </span></span>
<span><span class="co">##   The .summary column was added to the database.</span></span></code></pre></div>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save the uncleaned dataset</span></span>
<span><span class="va">check_time</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"05_unCleaned_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter">8.3 Filter<a class="anchor" aria-label="anchor" href="#filter"></a>
</h3>
<p>Now clean the dataset of extra columns and failed rows and then save
it.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cleanData</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryFun.html">summaryFun</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">check_time</span>,</span>
<span>  dontFilterThese <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gridSummary"</span>, <span class="st">".lonFlag"</span>, <span class="st">".latFlag"</span>, <span class="st">".uncer_terms"</span>,</span>
<span>                      <span class="st">".uncertaintyThreshold"</span><span class="op">)</span>,</span>
<span>  <span class="co"># Remove the filtering columns?</span></span>
<span>  removeFilterColumns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="co"># Filter to ONLY cleaned data?</span></span>
<span>  filterClean <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  - We will NOT flag the following columns. However, they will remain in the data file.</span></span>
<span><span class="co">## .gridSummary, .lonFlag, .latFlag, .uncer_terms, .uncertaintyThreshold</span></span>
<span><span class="co">##  - summaryFun:</span></span>
<span><span class="co">## Flagged 93 </span></span>
<span><span class="co">##   The .summary column was added to the database.</span></span>
<span><span class="co">##  - REMOVED all occurrences that were FALSE for the 'summary' column.</span></span></code></pre></div>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save this CLEANED dataset</span></span>
<span><span class="va">cleanData</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_excel_csv</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"05_cleaned_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="figures-and-tables">9.0 Figures and tables<a class="anchor" aria-label="anchor" href="#figures-and-tables"></a>
</h2>
<div class="section level3">
<h3 id="duplicate-chorddiagrams">9.1 Duplicate chordDiagrams<a class="anchor" aria-label="anchor" href="#duplicate-chorddiagrams"></a>
</h3>
<p>Install <strong>BiocManager</strong> and
<strong>ComplexHeatmap</strong> if you missed them at the start.</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span><span class="op">)</span></span></code></pre></div>
<p>Read in the most recent file of flagged duplicates, if it’s not
already in your environment.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"duplicates"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">duplicates</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/fileFinder.html">fileFinder</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">DataPath</span>, fileName <span class="op">=</span> <span class="st">"duplicateRun_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Choose the global figure parameters.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Create the chordDiagram. You can leave many of the below values out,
but we show here the defaults. There are [internally] no duplicates in
current our test dataset, so <strong>BeeBDC</strong> will throw an
informative error. However, we show the full output figure from our bee
dataset below.</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/chordDiagramR.html">chordDiagramR</a></span><span class="op">(</span></span>
<span>  <span class="co"># The duplicate data from the dupeSummary function output  </span></span>
<span>  dupeData <span class="op">=</span> <span class="va">duplicates</span>,</span>
<span>  outPath <span class="op">=</span> <span class="va">OutPath_Figures</span>,</span>
<span>  fileName <span class="op">=</span> <span class="st">"ChordDiagram.pdf"</span>,</span>
<span>  <span class="co"># These can be modified to help fit the final pdf that's exported.</span></span>
<span>  width <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">7.5</span>,</span>
<span>  bg <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  <span class="co"># How few distinct dataSources should a group have to be listed as "other"</span></span>
<span>  smallGrpThreshold <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Duplicated record sources"</span>,</span>
<span>  <span class="co"># The default list of colour palettes to choose from usign the paleteer package</span></span>
<span>  palettes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cartography::blue.pal"</span>, <span class="st">"cartography::green.pal"</span>, </span>
<span>               <span class="st">"cartography::sand.pal"</span>, <span class="st">"cartography::orange.pal"</span>, <span class="st">"cartography::red.pal"</span>,</span>
<span>               <span class="st">"cartography::purple.pal"</span>, <span class="st">"cartography::brown.pal"</span><span class="op">)</span>,</span>
<span>  canvas.ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.0</span>,<span class="fl">1.0</span><span class="op">)</span>, </span>
<span>  canvas.xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.6</span>, <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>  text.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  legendX <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  legendY <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">18</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  legendJustify <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span>,</span>
<span>  niceFacing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>![Full chord diagram from Dorey et al. 2023]
<img src="https://photos.smugmug.com/photos/i-Xt3tN8L/0/X5/i-Xt3tN8L-X5.jpg" align="left"></p>
</div>
<div class="section level3">
<h3 id="duplicate-histogram">9.2 Duplicate histogram<a class="anchor" aria-label="anchor" href="#duplicate-histogram"></a>
</h3>
<p>Read in the uncleaned dataset, if it’s not already present.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"check_time"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">beeData</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"05_unCleaned_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">beeData</span> <span class="op">&lt;-</span> <span class="va">check_time</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">check_time</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Create a plot with two bar graphs. One shows the absolute number of
duplicate records for each data source, while the other shows the
proportion of records that are duplicated within each data source.
(<em>‘dataSource’</em> is simplified to the text before the first
underscore).</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dupePlotR.html">dupePlotR</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">beeData</span>,</span>
<span>  <span class="co"># The outPath to save the plot as</span></span>
<span>  outPath <span class="op">=</span> <span class="va">OutPath_Figures</span>,</span>
<span>  fileName <span class="op">=</span> <span class="st">"duplicatePlot.pdf"</span>,</span>
<span>  <span class="co"># Colours in order: duplicate, kept duplicate, unique</span></span>
<span>  dupeColours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F2D2A2"</span>,<span class="st">"#B9D6BC"</span>, <span class="st">"#349B90"</span><span class="op">)</span>,</span>
<span>  <span class="co"># Plot size and height</span></span>
<span>  base_height <span class="op">=</span> <span class="fl">7</span>, base_width <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.85</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>  <span class="co"># Extra variables can be fed into forcats::fct_recode() to change names on plot</span></span>
<span>  GBIF <span class="op">=</span> <span class="st">"GBIF"</span>, SCAN <span class="op">=</span> <span class="st">"SCAN"</span>, iDigBio <span class="op">=</span> <span class="st">"iDigBio"</span>, USGS <span class="op">=</span> <span class="st">"USGS"</span>, ALA <span class="op">=</span> <span class="st">"ALA"</span>, </span>
<span>  ASP <span class="op">=</span> <span class="st">"ASP"</span>, CAES <span class="op">=</span> <span class="st">"CAES"</span>, Ecd <span class="op">=</span> <span class="st">"Ecd"</span>,</span>
<span>  returnPlot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="BeeBDC_main_files/figure-html/9.2ii-1.png" width="700"></p>
<p>![Full duplicate diagram from Dorey et al. 2023]
<img src="https://photos.smugmug.com/photos/i-CjrvS3d/0/NWnPWxvP8VGmTwZKjtJbgc5xCzjJwCt4P2PK7Rvqz/X4/i-CjrvS3d-X4.jpg" align="left"></p>
</div>
<div class="section level3">
<h3 id="flags-by-source">9.3 Flags by source<a class="anchor" aria-label="anchor" href="#flags-by-source"></a>
</h3>
<p>Create a compound bar plot that shows the proportion of records that
pass or fail each flag (rows) for each data source (columns). The
function can also optionally return a point map for a user-specified
species when plotMap = TRUE. (<em>dataSource</em> is simplified to the
text before the first underscore.)</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/plotFlagSummary.html">plotFlagSummary</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">beeData</span>,</span>
<span>  <span class="co"># Colours in order of pass (TRUE), fail (FALSE), and NA</span></span>
<span>  flagColours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#127852"</span>, <span class="st">"#A7002D"</span>, <span class="st">"#BDBABB"</span><span class="op">)</span>,</span>
<span>  fileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"FlagsPlot_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">".pdf"</span><span class="op">)</span>,</span>
<span>  outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">OutPath_Figures</span><span class="op">)</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">15</span>, height <span class="op">=</span> <span class="fl">9</span>,</span>
<span>    <span class="co"># OPTIONAL:</span></span>
<span>      <span class="co">#   # Filter to a single species</span></span>
<span>      <span class="co">#       speciesName = "Holcopasites heliopsis",</span></span>
<span>      <span class="co">#         # column to look in</span></span>
<span>      <span class="co">#       nameColumn = "species",</span></span>
<span>      <span class="co">#        # Save the filtered data</span></span>
<span>      <span class="co">#       saveFiltered = TRUE,</span></span>
<span>      <span class="co">#   # Filter column to display on map</span></span>
<span>      <span class="co">#       filterColumn = ".summary",</span></span>
<span>      <span class="co">#       plotMap = TRUE,</span></span>
<span>      <span class="co">#   # amount to jitter points if desired, e.g. 0.25 or NULL</span></span>
<span>      <span class="co">#       jitterValue = NULL,</span></span>
<span>      <span class="co">#        # Map opacity value for points between 0 and 1</span></span>
<span>      <span class="co">#   mapAlpha = 1,</span></span>
<span>      <span class="co">#        # If a user wants to output the table used to make the figure, change this to TRUE</span></span>
<span>      <span class="co">#   saveTable = FALSE,</span></span>
<span>  <span class="co"># Extra variables can be fed into forcats::fct_recode() to change names on plot</span></span>
<span>  GBIF <span class="op">=</span> <span class="st">"GBIF"</span>, SCAN <span class="op">=</span> <span class="st">"SCAN"</span>, iDigBio <span class="op">=</span> <span class="st">"iDigBio"</span>, USGS <span class="op">=</span> <span class="st">"USGS"</span>, ALA <span class="op">=</span> <span class="st">"ALA"</span>, </span>
<span>  ASP <span class="op">=</span> <span class="st">"ASP"</span>, CAES <span class="op">=</span> <span class="st">"CAES"</span>, <span class="st">'BMont'</span> <span class="op">=</span> <span class="st">"BMont"</span>, <span class="st">'BMin'</span> <span class="op">=</span> <span class="st">"BMin"</span>, Ecd <span class="op">=</span> <span class="st">"Ecd"</span>,</span>
<span>  Gaiarsa <span class="op">=</span> <span class="st">"Gai"</span>, EPEL <span class="op">=</span> <span class="st">"EPEL"</span>, VicWam <span class="op">=</span> <span class="st">"VicWam"</span>,</span>
<span>  returnPlot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##  - Preparing data to plot...</span></span>
<span><span class="co">##  - Building plot...</span></span></code></pre></div>
<p><img src="BeeBDC_main_files/figure-html/9.3-1.png" width="1440"></p>
</div>
<div class="section level3">
<h3 id="maps">9.4 Maps<a class="anchor" aria-label="anchor" href="#maps"></a>
</h3>
<p>Import CLEANED dataset (you can change this option).</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"cleanData"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cleanData</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"05_cleaned_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level4">
<h4 id="a--summary-maps">a. Summary maps<a class="anchor" aria-label="anchor" href="#a--summary-maps"></a>
</h4>
<p>Draw a global summary map for occurrence and species number by
country.</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/summaryMaps.html">summaryMaps</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">cleanData</span>, width <span class="op">=</span> <span class="fl">10</span>, height <span class="op">=</span> <span class="fl">10</span>, class_n <span class="op">=</span> <span class="fl">3</span>, class_Style <span class="op">=</span> <span class="st">"fisher"</span>,</span>
<span>    fileName <span class="op">=</span> <span class="st">"CountryMaps_fisher.pdf"</span>, outPath <span class="op">=</span> <span class="va">OutPath_Figures</span>, returnPlot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="b--interactive-maps">b. Interactive maps<a class="anchor" aria-label="anchor" href="#b--interactive-maps"></a>
</h4>
<p>Uses the occurrence data (preferably uncleaned in order to show
pass/fail points) and outputs interactive .html maps, which can be
opened in your browser, to a specific directory. The maps can highlight
if an occurrence has passed all filtering (<em>.summary</em> == TRUE) or
failed at least one filter (<em>.summary</em> == FALSE). This can be
updated by first running <code><a href="../reference/summaryFun.html">BeeBDC::summaryFun()</a></code> to choose the
columns that you want to be highlighted. This function will also
highlight occurrences flagged as expert-identified or country outliers
separately. Because the function can have any categorical variable fed
into ‘speciesColumn’, users may choose another column of interest to
map; however, maps made using very large categories can be slow to
produce and unwieldy to view.</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/interactiveMapR.html">interactiveMapR</a></span><span class="op">(</span></span>
<span>   <span class="co"># occurrence data</span></span>
<span>  data <span class="op">=</span> <span class="va">beeData</span>,</span>
<span>   <span class="co"># Directory where to save files</span></span>
<span>  outPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">OutPath_Figures</span>, <span class="st">"interactiveMaps"</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>,</span>
<span>  lon <span class="op">=</span> <span class="st">"decimalLongitude"</span>,</span>
<span>  lat <span class="op">=</span> <span class="st">"decimalLatitude"</span>,</span>
<span>    <span class="co"># Occurrence dataset column with species names</span></span>
<span>  speciesColumn <span class="op">=</span> <span class="st">"scientificName"</span>,</span>
<span>    <span class="co"># Which species to map — a character vector of names or "ALL"</span></span>
<span>    <span class="co"># Note: "ALL" is defined AFTER filtering for country</span></span>
<span>  speciesList <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>  countryList <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># study area</span></span>
<span>    <span class="co"># Point jitter to see stacked points — jitters an amount in decimal degrees</span></span>
<span>  jitterValue <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="data-providers">9.5 Data providers<a class="anchor" aria-label="anchor" href="#data-providers"></a>
</h3>
<p>Read in the clean data if it’s not already in the environment.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/exists.html" class="external-link">exists</a></span><span class="op">(</span><span class="st">"cleanData"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cleanData</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">OutPath_Intermediate</span>, <span class="st">"05_cleaned_database.csv"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/ColTypeR.html">ColTypeR</a></span><span class="op">(</span><span class="op">)</span>, locale <span class="op">=</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/locale.html" class="external-link">locale</a></span><span class="op">(</span>encoding <span class="op">=</span> <span class="st">"UTF-8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>This function will attempt to find and build a table of data
providers that have contributed to the input data, especially using the
<em>‘institutionCode’</em> column. It will also search a variety of
other columns to find data providers using an internally set sequence of
if-else statements. Hence, this function is quite specific for bee data,
but it should work for other taxa in similar institutions (perhaps to a
lesser degree).</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note, if outPath = NULL then no file will be saved</span></span>
<span><span class="va">dataProvTable</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/dataProvTables.html">dataProvTables</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">cleanData</span>, runBeeDataChecks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    outPath <span class="op">=</span> <span class="cn">NULL</span>, fileName <span class="op">=</span> <span class="st">"dataProvTable.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="flag-summary">9.6 Flag summary<a class="anchor" aria-label="anchor" href="#flag-summary"></a>
</h3>
<p>The function <code><a href="../reference/flagSummaryTable.html">BeeBDC::flagSummaryTable()</a></code> takes a flagged
dataset and returns the total number of fails (FALSE) per flag (in
columns starting with “.”) and per species. Users may define the column
by which to group the summary. While it is intended to work with the
<em>scientificName</em> column, users may select any grouping column
(e.g., <em>country</em>).</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note, if outPath = NULL then no file will be saved</span></span>
<span><span class="va">summaryTable</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="../reference/flagSummaryTable.html">flagSummaryTable</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">beeData</span>, column <span class="op">=</span> <span class="st">"scientificName"</span>,</span>
<span>    outPath <span class="op">=</span> <span class="cn">NULL</span>, fileName <span class="op">=</span> <span class="st">"flagTable.csv"</span><span class="op">)</span></span>
<span><span class="co">##  - We will flag all columns starting with '.'</span></span>
<span><span class="co">##  - summaryFun:</span></span>
<span><span class="co">## Flagged 107 </span></span>
<span><span class="co">##   The .summary column was added to the database.</span></span>
<span><span class="co">## The percentages of species impacted by each flag in your analysis are as follows: </span></span>
<span><span class="co">##   .coordinates_empty = 24.05%</span></span>
<span><span class="co">##   .coordinates_outOfRange = 0%</span></span>
<span><span class="co">##   .basisOfRecords_notStandard = 1.27%</span></span>
<span><span class="co">##   .coordinates_country_inconsistent = 1.27%</span></span>
<span><span class="co">##   .occurrenceAbsent = 8.86%</span></span>
<span><span class="co">##   .unLicensed = 0%</span></span>
<span><span class="co">##   .GBIFflags = 0%</span></span>
<span><span class="co">##   .rou = 30.38%</span></span>
<span><span class="co">##   .sequential = 0%</span></span>
<span><span class="co">##   .uncertaintyThreshold = 15.19%</span></span>
<span><span class="co">##   .eventDate_empty = 16.46%</span></span>
<span><span class="co">##   .year_outOfRange = 16.46%</span></span>
<span><span class="co">##   .duplicates = 0%</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey, Robert L. O’Reilly, Silas Bossert, Erica E. Fischer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
